// @xlabs-libs/wallet-aggregator-core/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_ID_SEDA = exports.CHAIN_ID_STARGAZE = exports.CHAIN_ID_CELESTIA = exports.CHAIN_ID_NEUTRON = exports.CHAIN_ID_KUJIRA = exports.CHAIN_ID_EVMOS = exports.CHAIN_ID_COSMOSHUB = exports.CHAIN_ID_WORMCHAIN = exports.CHAIN_ID_SEIEVM = exports.CHAIN_ID_BERACHAIN = exports.CHAIN_ID_LINEA = exports.CHAIN_ID_WORLDCHAIN = exports.CHAIN_ID_XLAYER = exports.CHAIN_ID_BLAST = exports.CHAIN_ID_MANTLE = exports.CHAIN_ID_SCROLL = exports.CHAIN_ID_ROOTSTOCK = exports.CHAIN_ID_SEI = exports.CHAIN_ID_BASE = exports.CHAIN_ID_BTC = exports.CHAIN_ID_XPLA = exports.CHAIN_ID_PYTHNET = exports.CHAIN_ID_GNOSIS = exports.CHAIN_ID_OPTIMISM = exports.CHAIN_ID_ARBITRUM = exports.CHAIN_ID_APTOS = exports.CHAIN_ID_SUI = exports.CHAIN_ID_OSMOSIS = exports.CHAIN_ID_INJECTIVE = exports.CHAIN_ID_TERRA2 = exports.CHAIN_ID_NEON = exports.CHAIN_ID_MOONBEAM = exports.CHAIN_ID_NEAR = exports.CHAIN_ID_CELO = exports.CHAIN_ID_KLAYTN = exports.CHAIN_ID_ACALA = exports.CHAIN_ID_KARURA = exports.CHAIN_ID_FANTOM = exports.CHAIN_ID_AURORA = exports.CHAIN_ID_ALGORAND = exports.CHAIN_ID_OASIS = exports.CHAIN_ID_AVAX = exports.CHAIN_ID_POLYGON = exports.CHAIN_ID_BSC = exports.CHAIN_ID_TERRA = exports.CHAIN_ID_ETH = exports.CHAIN_ID_SOLANA = exports.CHAIN_ID_UNSET = exports.CHAIN_ID_TO_NAME = exports.CHAINS = void 0;
exports.isTerraChain = exports.isCosmWasmChain = exports.isEVMChain = exports.coalesceChainName = exports.coalesceChainId = exports.toChainName = exports.toChainId = exports.isChain = exports.CHAIN_ID_MONAD = exports.CHAIN_ID_UNICHAIN = exports.CHAIN_ID_POLYGON_SEPOLIA = exports.CHAIN_ID_HOLESKY = exports.CHAIN_ID_OPTIMISM_SEPOLIA = exports.CHAIN_ID_BASE_SEPOLIA = exports.CHAIN_ID_ARBITRUM_SEPOLIA = exports.CHAIN_ID_SEPOLIA = exports.CHAIN_ID_PROVENANCE = exports.CHAIN_ID_DYMENSION = void 0;
// From: https://github.com/wormhole-foundation/wormhole/blob/dev.v2/sdk/js/src/utils/consts.ts#L1
exports.CHAINS = {
    unset: 0,
    solana: 1,
    ethereum: 2,
    terra: 3,
    bsc: 4,
    polygon: 5,
    avalanche: 6,
    oasis: 7,
    algorand: 8,
    aurora: 9,
    fantom: 10,
    karura: 11,
    acala: 12,
    klaytn: 13,
    celo: 14,
    near: 15,
    moonbeam: 16,
    neon: 17,
    terra2: 18,
    injective: 19,
    osmosis: 20,
    sui: 21,
    aptos: 22,
    arbitrum: 23,
    optimism: 24,
    gnosis: 25,
    pythnet: 26,
    xpla: 28,
    btc: 29,
    base: 30,
    sei: 32,
    rootstock: 33, //
    scroll: 34,
    mantle: 35,
    blast: 36,
    xlayer: 37,
    linea: 38,
    berachain: 39,
    seievm: 40,
    unichain: 44,
    worldchain: 45,
    monad: 48,
    wormchain: 3104,
    cosmoshub: 4000,
    evmos: 4001,
    kujira: 4002,
    neutron: 4003,
    celestia: 4004,
    stargaze: 4005,
    seda: 4006,
    dymension: 4007,
    provenance: 4008,
    sepolia: 10002,
    arbitrum_sepolia: 10003,
    base_sepolia: 10004,
    optimism_sepolia: 10005,
    holesky: 10006,
    polygon_sepolia: 10007,
};
exports.CHAIN_ID_TO_NAME = Object.entries(exports.CHAINS).reduce((obj, [name, id]) => {
    obj[id] = name; // eslint-disable-line
    return obj; // eslint-disable-line
}, {} // eslint-disable-line
);
exports.CHAIN_ID_UNSET = exports.CHAINS["unset"];
exports.CHAIN_ID_SOLANA = exports.CHAINS["solana"];
exports.CHAIN_ID_ETH = exports.CHAINS["ethereum"];
exports.CHAIN_ID_TERRA = exports.CHAINS["terra"];
exports.CHAIN_ID_BSC = exports.CHAINS["bsc"];
exports.CHAIN_ID_POLYGON = exports.CHAINS["polygon"];
exports.CHAIN_ID_AVAX = exports.CHAINS["avalanche"];
exports.CHAIN_ID_OASIS = exports.CHAINS["oasis"];
exports.CHAIN_ID_ALGORAND = exports.CHAINS["algorand"];
exports.CHAIN_ID_AURORA = exports.CHAINS["aurora"];
exports.CHAIN_ID_FANTOM = exports.CHAINS["fantom"];
exports.CHAIN_ID_KARURA = exports.CHAINS["karura"];
exports.CHAIN_ID_ACALA = exports.CHAINS["acala"];
exports.CHAIN_ID_KLAYTN = exports.CHAINS["klaytn"];
exports.CHAIN_ID_CELO = exports.CHAINS["celo"];
exports.CHAIN_ID_NEAR = exports.CHAINS["near"];
exports.CHAIN_ID_MOONBEAM = exports.CHAINS["moonbeam"];
exports.CHAIN_ID_NEON = exports.CHAINS["neon"];
exports.CHAIN_ID_TERRA2 = exports.CHAINS["terra2"];
exports.CHAIN_ID_INJECTIVE = exports.CHAINS["injective"];
exports.CHAIN_ID_OSMOSIS = exports.CHAINS["osmosis"];
exports.CHAIN_ID_SUI = exports.CHAINS["sui"];
exports.CHAIN_ID_APTOS = exports.CHAINS["aptos"];
exports.CHAIN_ID_ARBITRUM = exports.CHAINS["arbitrum"];
exports.CHAIN_ID_OPTIMISM = exports.CHAINS["optimism"];
exports.CHAIN_ID_GNOSIS = exports.CHAINS["gnosis"];
exports.CHAIN_ID_PYTHNET = exports.CHAINS["pythnet"];
exports.CHAIN_ID_XPLA = exports.CHAINS["xpla"];
exports.CHAIN_ID_BTC = exports.CHAINS["btc"];
exports.CHAIN_ID_BASE = exports.CHAINS["base"];
exports.CHAIN_ID_SEI = exports.CHAINS["sei"];
exports.CHAIN_ID_ROOTSTOCK = exports.CHAINS["rootstock"];
exports.CHAIN_ID_SCROLL = exports.CHAINS["scroll"];
exports.CHAIN_ID_MANTLE = exports.CHAINS["mantle"];
exports.CHAIN_ID_BLAST = exports.CHAINS["blast"];
exports.CHAIN_ID_XLAYER = exports.CHAINS["xlayer"];
exports.CHAIN_ID_WORLDCHAIN = exports.CHAINS["worldchain"];
exports.CHAIN_ID_LINEA = exports.CHAINS["linea"];
exports.CHAIN_ID_BERACHAIN = exports.CHAINS["berachain"];
exports.CHAIN_ID_SEIEVM = exports.CHAINS["seievm"];
exports.CHAIN_ID_WORMCHAIN = exports.CHAINS["wormchain"];
exports.CHAIN_ID_COSMOSHUB = exports.CHAINS["cosmoshub"];
exports.CHAIN_ID_EVMOS = exports.CHAINS["evmos"];
exports.CHAIN_ID_KUJIRA = exports.CHAINS["kujira"];
exports.CHAIN_ID_NEUTRON = exports.CHAINS["neutron"];
exports.CHAIN_ID_CELESTIA = exports.CHAINS["celestia"];
exports.CHAIN_ID_STARGAZE = exports.CHAINS["stargaze"];
exports.CHAIN_ID_SEDA = exports.CHAINS["seda"];
exports.CHAIN_ID_DYMENSION = exports.CHAINS["dymension"];
exports.CHAIN_ID_PROVENANCE = exports.CHAINS["provenance"];
exports.CHAIN_ID_SEPOLIA = exports.CHAINS["sepolia"];
exports.CHAIN_ID_ARBITRUM_SEPOLIA = exports.CHAINS["arbitrum_sepolia"];
exports.CHAIN_ID_BASE_SEPOLIA = exports.CHAINS["base_sepolia"];
exports.CHAIN_ID_OPTIMISM_SEPOLIA = exports.CHAINS["optimism_sepolia"];
exports.CHAIN_ID_HOLESKY = exports.CHAINS["holesky"];
exports.CHAIN_ID_POLYGON_SEPOLIA = exports.CHAINS["polygon_sepolia"];
exports.CHAIN_ID_UNICHAIN = exports.CHAINS["unichain"];
exports.CHAIN_ID_MONAD = exports.CHAINS["monad"];
function isChain(chain) {
    if (typeof chain === "number") {
        return chain in exports.CHAIN_ID_TO_NAME;
    }
    else {
        return chain in exports.CHAINS;
    }
}
exports.isChain = isChain;
function toChainId(chainName) {
    return exports.CHAINS[chainName];
}
exports.toChainId = toChainId;
function toChainName(chainId) {
    return exports.CHAIN_ID_TO_NAME[chainId];
}
exports.toChainName = toChainName;
function coalesceChainId(chain) {
    return typeof chain === "number" && isChain(chain) ? chain : toChainId(chain);
}
exports.coalesceChainId = coalesceChainId;
function coalesceChainName(chain) {
    return toChainName(coalesceChainId(chain));
}
exports.coalesceChainName = coalesceChainName;
function isEVMChain(chainId) {
    return (chainId === exports.CHAIN_ID_ETH ||
        chainId === exports.CHAIN_ID_BSC ||
        chainId === exports.CHAIN_ID_AVAX ||
        chainId === exports.CHAIN_ID_POLYGON ||
        chainId === exports.CHAIN_ID_OASIS ||
        chainId === exports.CHAIN_ID_AURORA ||
        chainId === exports.CHAIN_ID_FANTOM ||
        chainId === exports.CHAIN_ID_KARURA ||
        chainId === exports.CHAIN_ID_ACALA ||
        chainId === exports.CHAIN_ID_KLAYTN ||
        chainId === exports.CHAIN_ID_CELO ||
        chainId === exports.CHAIN_ID_MOONBEAM ||
        chainId === exports.CHAIN_ID_NEON ||
        chainId === exports.CHAIN_ID_ARBITRUM ||
        chainId === exports.CHAIN_ID_OPTIMISM ||
        chainId === exports.CHAIN_ID_GNOSIS ||
        chainId === exports.CHAIN_ID_BASE ||
        chainId === exports.CHAIN_ID_SCROLL ||
        chainId === exports.CHAIN_ID_WORLDCHAIN ||
        chainId === exports.CHAIN_ID_BLAST ||
        chainId === exports.CHAIN_ID_XLAYER ||
        chainId === exports.CHAIN_ID_MANTLE ||
        chainId === exports.CHAIN_ID_UNICHAIN ||
        chainId === exports.CHAIN_ID_MONAD);
}
exports.isEVMChain = isEVMChain;
function isCosmWasmChain(chainId) {
    return (chainId === exports.CHAIN_ID_TERRA ||
        chainId === exports.CHAIN_ID_TERRA2 ||
        chainId === exports.CHAIN_ID_INJECTIVE ||
        chainId === exports.CHAIN_ID_XPLA ||
        chainId === exports.CHAIN_ID_SEI ||
        chainId === exports.CHAIN_ID_WORMCHAIN ||
        chainId === exports.CHAIN_ID_COSMOSHUB ||
        chainId === exports.CHAIN_ID_EVMOS ||
        chainId === exports.CHAIN_ID_KUJIRA);
}
exports.isCosmWasmChain = isCosmWasmChain;
function isTerraChain(chainId) {
    return chainId === exports.CHAIN_ID_TERRA || chainId === exports.CHAIN_ID_TERRA2;
}
exports.isTerraChain = isTerraChain;
