// @binance/w3w-qrcode-modal/src/browser/assets/baseStyle.ts
export const baseStyle = `#binanceW3W-wrapper :is(.pointer-events-auto) {
  pointer-events: auto;
}

#binanceW3W-wrapper :is(.fixed) {
  position: fixed;
}

#binanceW3W-wrapper :is(.absolute) {
  position: absolute;
}

#binanceW3W-wrapper :is(.relative) {
  position: relative;
}

#binanceW3W-wrapper :is(.bottom-0) {
  bottom: 0px;
}

#binanceW3W-wrapper :is(.left-0) {
  left: 0px;
}

#binanceW3W-wrapper :is(.top-0) {
  top: 0px;
}

#binanceW3W-wrapper :is(.m-auto) {
  margin: auto;
}

#binanceW3W-wrapper :is(.mx-\\[4px\\]) {
  margin-left: 4px;
  margin-right: 4px;
}

#binanceW3W-wrapper :is(.mb-4) {
  margin-bottom: 1rem;
}

#binanceW3W-wrapper :is(.mb-6) {
  margin-bottom: 1.5rem;
}

#binanceW3W-wrapper :is(.ml-1) {
  margin-left: 0.25rem;
}

#binanceW3W-wrapper :is(.mt-6) {
  margin-top: 1.5rem;
}

#binanceW3W-wrapper :is(.mt-\\[35px\\]) {
  margin-top: 35px;
}

#binanceW3W-wrapper :is(.flex) {
  display: flex;
}

#binanceW3W-wrapper :is(.grid) {
  display: grid;
}

#binanceW3W-wrapper :is(.h-\\[200px\\]) {
  height: 200px;
}

#binanceW3W-wrapper :is(.h-\\[24px\\]) {
  height: 24px;
}

#binanceW3W-wrapper :is(.h-\\[40px\\]) {
  height: 40px;
}

#binanceW3W-wrapper :is(.h-\\[52px\\]) {
  height: 52px;
}

#binanceW3W-wrapper :is(.h-full) {
  height: 100%;
}

#binanceW3W-wrapper :is(.w-\\[150px\\]) {
  width: 150px;
}

#binanceW3W-wrapper :is(.w-\\[200px\\]) {
  width: 200px;
}

#binanceW3W-wrapper :is(.w-\\[20px\\]) {
  width: 20px;
}

#binanceW3W-wrapper :is(.w-\\[24px\\]) {
  width: 24px;
}

#binanceW3W-wrapper :is(.w-\\[343px\\]) {
  width: 343px;
}

#binanceW3W-wrapper :is(.w-\\[60px\\]) {
  width: 60px;
}

#binanceW3W-wrapper :is(.w-\\[75px\\]) {
  width: 75px;
}

#binanceW3W-wrapper :is(.w-full) {
  width: 100%;
}

#binanceW3W-wrapper :is(.transform) {
  transform: translate(var(--tw-translate-x), var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
}

#binanceW3W-wrapper :is(.cursor-pointer) {
  cursor: pointer;
}

#binanceW3W-wrapper :is(.grid-flow-col) {
  grid-auto-flow: column;
}

#binanceW3W-wrapper :is(.grid-cols-2) {
  grid-template-columns: repeat(2, minmax(0, 1fr));
}

#binanceW3W-wrapper :is(.items-center) {
  align-items: center;
}

#binanceW3W-wrapper :is(.justify-end) {
  justify-content: flex-end;
}

#binanceW3W-wrapper :is(.justify-center) {
  justify-content: center;
}

#binanceW3W-wrapper :is(.justify-between) {
  justify-content: space-between;
}

#binanceW3W-wrapper :is(.gap-x-4) {
  -moz-column-gap: 1rem;
       column-gap: 1rem;
}

#binanceW3W-wrapper :is(.gap-y-2) {
  row-gap: 0.5rem;
}

#binanceW3W-wrapper :is(.rounded) {
  border-radius: 0.25rem;
}

#binanceW3W-wrapper :is(.rounded-2xl) {
  border-radius: 1rem;
}

#binanceW3W-wrapper :is(.rounded-lg) {
  border-radius: 0.5rem;
}

#binanceW3W-wrapper :is(.rounded-t-2xl) {
  border-top-left-radius: 1rem;
  border-top-right-radius: 1rem;
}

#binanceW3W-wrapper :is(.border) {
  border-width: 1px;
}

#binanceW3W-wrapper :is(.border-b) {
  border-bottom-width: 1px;
}

#binanceW3W-wrapper :is(.border-gray-300) {
  --tw-border-opacity: 1;
  border-color: rgb(209 213 219 / var(--tw-border-opacity));
}

#binanceW3W-wrapper :is(.bg-black\\/\\[\\.80\\]) {
  background-color: rgb(0 0 0 / .80);
}

#binanceW3W-wrapper :is(.bg-white) {
  --tw-bg-opacity: 1;
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
}

#binanceW3W-wrapper :is(.p-\\[12px\\]) {
  padding: 12px;
}

#binanceW3W-wrapper :is(.px-4) {
  padding-left: 1rem;
  padding-right: 1rem;
}

#binanceW3W-wrapper :is(.px-6) {
  padding-left: 1.5rem;
  padding-right: 1.5rem;
}

#binanceW3W-wrapper :is(.py-3) {
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
}

#binanceW3W-wrapper :is(.py-4) {
  padding-top: 1rem;
  padding-bottom: 1rem;
}

#binanceW3W-wrapper :is(.pb-6) {
  padding-bottom: 1.5rem;
}

#binanceW3W-wrapper :is(.pt-\\[20px\\]) {
  padding-top: 20px;
}

#binanceW3W-wrapper :is(.text-center) {
  text-align: center;
}

#binanceW3W-wrapper :is(.text-base) {
  font-size: 1rem;
  line-height: 1.5rem;
}

#binanceW3W-wrapper :is(.font-medium) {
  font-weight: 500;
}

#binanceW3W-wrapper :is(.text-\\[\\#1E2329\\]) {
  --tw-text-opacity: 1;
  color: rgb(30 35 41 / var(--tw-text-opacity));
}

#binanceW3W-wrapper :is(.text-\\[\\#929AA5\\]) {
  --tw-text-opacity: 1;
  color: rgb(146 154 165 / var(--tw-text-opacity));
}

#binanceW3W-wrapper :is(.shadow-inner) {
  --tw-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 0.05);
  --tw-shadow-colored: inset 0 2px 4px 0 var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow);
}

#binanceW3W-wrapper :is(.duration-300) {
  transition-duration: 300ms;
}

#binanceW3W-wrapper :is(.ease-in-out) {
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
}

#binanceW3W-wrapper :is(.will-change-auto) {
  will-change: auto;
}

#binanceW3W-wrapper :is(.will-change-transform) {
  will-change: transform;
}

.w3w-body3 {
  font-size: 14px;
  font-style: normal;
  font-weight: 400;
  line-height: 20px;
}

.w3w-subtitle1 {
  font-size: 20px;
  font-style: normal;
  font-weight: 600;
  line-height: 28px;
}

.w3w-subtitle3 {
  font-size: 16px;
  font-style: normal;
  font-weight: 500;
  line-height: 24px;
  /* 150% */
}

.w3w-caption2 {
  font-size: 12px;
  font-style: normal;
  font-weight: 400;
  line-height: 16px;
}

.w3w-t-black {
  color: #0b0e11;
}

.w3w-t-brand {
  color: #c99400;
}

.w3w-t-primary {
  color: #202630;
}

.w3w-t-secondary {
  color: #474d57;
}

.w3w-bg-primary {
  background: #fcd535;
}

@media (min-width: 768px) {
  .md\\:w3w-subtitle1 {
    font-size: 20px;
    font-style: normal;
    font-weight: 600;
    line-height: 28px;
  }

  #binanceW3W-wrapper :is(.md\\:w-\\[400px\\]) {
    width: 400px;
  }

  #binanceW3W-wrapper :is(.md\\:font-semibold) {
    font-weight: 600;
  }
}

@media (min-width: 1024px) {
  #binanceW3W-wrapper :is(.lg\\:p-8) {
    padding: 2rem;
  }

  #binanceW3W-wrapper :is(.lg\\:pt-6) {
    padding-top: 1.5rem;
  }

  #binanceW3W-wrapper :is(.lg\\:text-xl) {
    font-size: 1.25rem;
    line-height: 1.75rem;
  }
}
`
