// @xlabs-libs/wallet-aggregator-evm/dist/esm/chains.js
import * as CHAINS from "@wagmi/core/chains";
export * from "@wagmi/core/chains";
const acala = {
    id: 787,
    name: "Acala",
    network: "acala",
    nativeCurrency: { name: "Acala Token", symbol: "ACA", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://eth-rpc-acala.aca-api.network"] },
        public: { http: ["https://eth-rpc-acala.aca-api.network"] },
    },
    blockExplorers: {
        default: { name: "Blockscout", url: "https://blockscout.acala.network" },
    },
};
const acalaTestnet = {
    id: 597,
    name: "Acala Testnet",
    network: "acala-testnet",
    nativeCurrency: { name: "Acala Token", symbol: "ACA", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://acala-dev.aca-dev.network/eth/http"] },
        public: { http: ["https://acala-dev.aca-dev.network/eth/http"] },
    },
    blockExplorers: {
        default: {
            name: "Blockscout",
            url: "https://blockscout.acala-dev.aca-dev.network",
        },
    },
};
const karura = {
    id: 686,
    name: "Karura",
    network: "karura",
    nativeCurrency: { name: "Karura Token", symbol: "KAR", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://eth-rpc-karura.aca-api.network"] },
        public: { http: ["https://eth-rpc-karura.aca-api.network"] },
    },
    blockExplorers: {
        default: { name: "Blockscout", url: "https://blockscout.karura.network" },
    },
};
const karuraTestnet = {
    id: 596,
    name: "Karura Testnet",
    network: "karura-testnet",
    nativeCurrency: { name: "Karura Token", symbol: "KAR", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://karura-dev.aca-dev.network/eth/http"] },
        public: { http: ["https://karura-dev.aca-dev.network/eth/http"] },
    },
    blockExplorers: {
        default: {
            name: "Blockscout",
            url: "https://blockscout.karura-dev.aca-dev.network",
        },
    },
};
export const emerald = {
    id: 42262,
    name: "Emerald Paratime Mainnet",
    network: "emerald-testnet",
    nativeCurrency: { name: "Emerald Rose", symbol: "ROSE", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://emerald.oasis.dev"] },
        public: { http: ["https://emerald.oasis.dev"] },
    },
    blockExplorers: {
        default: { name: "Oasis", url: "https://explorer.emerald.oasis.dev" },
    },
};
export const emeraldTestnet = {
    id: 42261,
    name: "Emerald Paratime Testnet",
    network: "emerald-testnet",
    nativeCurrency: { name: "Emerald Rose", symbol: "ROSE", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://testnet.emerald.oasis.dev"] },
        public: { http: ["https://testnet.emerald.oasis.dev"] },
    },
    blockExplorers: {
        default: {
            name: "Oasis",
            url: "https://testnet.explorer.emerald.oasis.dev",
        },
    },
};
// https://docs.blast.io/building/network-information
export const blast = {
    id: 81457,
    name: "Blast Mainnet",
    network: "blast-mainnet",
    nativeCurrency: { name: "Blast Ethereum", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://rpc.blast.io"] },
        public: { http: ["	https://rpc.blast.io"] },
    },
    blockExplorers: {
        default: {
            name: "Blast",
            url: "https://blastscan.io",
        },
    },
};
export const blastSepolia = {
    id: 168587773,
    name: "Blast Sepolia",
    network: "blast-repolia",
    nativeCurrency: { name: "Blast Ethereum", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://sepolia.blast.io"] },
        public: { http: ["https://sepolia.blast.io"] },
    },
    blockExplorers: {
        default: {
            name: "Blast Sepolia Explorer",
            url: "https://sepolia.blastexplorer.io",
        },
    },
};
export const scroll = {
    id: 534352,
    name: "Scroll",
    network: "scroll",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.ankr.com/scroll"],
        },
        public: {
            http: ["https://rpc.ankr.com/scroll"],
        },
    },
    blockExplorers: {
        default: {
            name: "Scrollscan",
            url: "https://scrollscan.com",
        },
    },
};
export const scrollSepolia = {
    id: 534351,
    name: "Scroll Sepolia",
    network: "scroll-sepolia",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.ankr.com/scroll_sepolia_testnet"],
        },
        public: {
            http: ["https://rpc.ankr.com/scroll_sepolia_testnet"],
        },
    },
    blockExplorers: {
        default: {
            name: "Scrollscan",
            url: "https://sepolia.scrollscan.com",
        },
    },
    testnet: true,
};
export const xlayer = {
    id: 196,
    name: "X Layer",
    network: "xlayer-mainnet",
    nativeCurrency: { name: "OKB", symbol: "OKB", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.xlayer.tech"],
        },
        public: {
            http: ["https://rpc.xlayer.tech"],
        },
    },
    blockExplorers: {
        default: {
            name: "X Layer Explorer",
            url: "https://www.okx.com/web3/explorer/xlayer",
        },
    },
};
export const xlayerTestnet = {
    id: 195,
    name: "X Layer Testnet",
    network: "xlayer-testnet",
    nativeCurrency: { name: "OKB", symbol: "OKB", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://testrpc.xlayer.tech"],
        },
        public: {
            http: ["https://testrpc.xlayer.tech"],
        },
    },
    blockExplorers: {
        default: {
            name: "X Layer Testnet Explorer",
            url: "https://www.okx.com/web3/explorer/xlayer-test",
        },
    },
    testnet: true,
};
export const mantle = {
    id: 5000,
    name: "Mantle",
    network: "mantle",
    nativeCurrency: { name: "MNT", symbol: "MNT", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.mantle.xyz"],
        },
        public: {
            http: ["https://rpc.mantle.xyz"],
        },
    },
    blockExplorers: {
        default: {
            name: "Mantle Explorer",
            url: "https://explorer.mantle.xyz",
        },
    },
};
export const mantleTestnet = {
    id: 5000,
    name: "Mantle",
    network: "mantle-testnet",
    nativeCurrency: { name: "MNT", symbol: "MNT", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.testnet.mantle.xyz"],
        },
        public: {
            http: ["https://rpc.testnet.mantle.xyz"],
        },
    },
    blockExplorers: {
        default: {
            name: "Mantle Explorer",
            url: "https://explorer.testnet.mantle.xyz",
        },
    },
};
export const worldchain = {
    id: 480,
    name: "World Chain",
    network: "worldchain",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://worldchain-mainnet.g.alchemy.com/public"],
        },
        public: {
            http: ["https://worldchain-mainnet.g.alchemy.com/public"],
        },
    },
    blockExplorers: {
        default: {
            name: "World Scan",
            url: "https://worldscan.org/",
        },
    },
};
export const worldchainTestnet = {
    id: 4801,
    name: "World Chain",
    network: "worldchain-sepolia",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://worldchain-sepolia.g.alchemy.com/public"],
        },
        public: {
            http: ["https://worldchain-sepolia.g.alchemy.com/public"],
        },
    },
    blockExplorers: {
        default: {
            name: "World Scan",
            url: "https://worldchain-sepolia.explorer.alchemy.com/",
        },
    },
    testnet: true,
};
export const unichain = {
    id: 130,
    name: "Unichain",
    network: "unichain",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://mainnet.unichain.org"],
        },
        public: {
            http: ["https://mainnet.unichain.org"],
        },
    },
    blockExplorers: {
        default: {
            name: "Unichain Explorer",
            url: "https://unichain.blockscout.com/",
        },
    },
};
export const unichainTestnet = {
    id: 1301,
    name: "Unichain Sepolia",
    network: "unichain-sepolia",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://sepolia.unichain.org"],
        },
        public: {
            http: ["https://sepolia.unichain.org"],
        },
    },
    blockExplorers: {
        default: {
            name: "Unichain Explorer",
            url: "https://unichain-sepolia.blockscout.com/",
        },
    },
    testnet: true,
};
export const monadTestnet = {
    id: 10143,
    name: "Monad Testnet",
    network: "monad-testnet",
    nativeCurrency: { name: "MON", symbol: "MON", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://testnet-rpc.monad.xyz"],
        },
        public: {
            http: ["https://testnet-rpc.monad.xyz"],
        },
    },
    blockExplorers: {
        default: {
            name: "Oasis",
            url: "https://testnet.explorer.oasis.dev",
        },
    },
    testnet: true,
};
export const DEFAULT_CHAINS = [
    ...Object.values(CHAINS),
    acala,
    acalaTestnet,
    blast,
    blastSepolia,
    emerald,
    emeraldTestnet,
    karura,
    karuraTestnet,
    scroll,
    scrollSepolia,
    xlayer,
    xlayerTestnet,
    mantle,
    mantleTestnet,
    worldchain,
    worldchainTestnet,
    unichain,
    unichainTestnet,
    monadTestnet,
];
