// @xlabs-libs/wallet-aggregator-evm/dist/cjs/chains.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CHAINS = exports.monadTestnet = exports.unichainTestnet = exports.unichain = exports.worldchainTestnet = exports.worldchain = exports.mantleTestnet = exports.mantle = exports.xlayerTestnet = exports.xlayer = exports.scrollSepolia = exports.scroll = exports.blastSepolia = exports.blast = exports.emeraldTestnet = exports.emerald = void 0;
const CHAINS = __importStar(require("@wagmi/core/chains"));
__exportStar(require("@wagmi/core/chains"), exports);
const acala = {
    id: 787,
    name: "Acala",
    network: "acala",
    nativeCurrency: { name: "Acala Token", symbol: "ACA", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://eth-rpc-acala.aca-api.network"] },
        public: { http: ["https://eth-rpc-acala.aca-api.network"] },
    },
    blockExplorers: {
        default: { name: "Blockscout", url: "https://blockscout.acala.network" },
    },
};
const acalaTestnet = {
    id: 597,
    name: "Acala Testnet",
    network: "acala-testnet",
    nativeCurrency: { name: "Acala Token", symbol: "ACA", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://acala-dev.aca-dev.network/eth/http"] },
        public: { http: ["https://acala-dev.aca-dev.network/eth/http"] },
    },
    blockExplorers: {
        default: {
            name: "Blockscout",
            url: "https://blockscout.acala-dev.aca-dev.network",
        },
    },
};
const karura = {
    id: 686,
    name: "Karura",
    network: "karura",
    nativeCurrency: { name: "Karura Token", symbol: "KAR", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://eth-rpc-karura.aca-api.network"] },
        public: { http: ["https://eth-rpc-karura.aca-api.network"] },
    },
    blockExplorers: {
        default: { name: "Blockscout", url: "https://blockscout.karura.network" },
    },
};
const karuraTestnet = {
    id: 596,
    name: "Karura Testnet",
    network: "karura-testnet",
    nativeCurrency: { name: "Karura Token", symbol: "KAR", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://karura-dev.aca-dev.network/eth/http"] },
        public: { http: ["https://karura-dev.aca-dev.network/eth/http"] },
    },
    blockExplorers: {
        default: {
            name: "Blockscout",
            url: "https://blockscout.karura-dev.aca-dev.network",
        },
    },
};
exports.emerald = {
    id: 42262,
    name: "Emerald Paratime Mainnet",
    network: "emerald-testnet",
    nativeCurrency: { name: "Emerald Rose", symbol: "ROSE", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://emerald.oasis.dev"] },
        public: { http: ["https://emerald.oasis.dev"] },
    },
    blockExplorers: {
        default: { name: "Oasis", url: "https://explorer.emerald.oasis.dev" },
    },
};
exports.emeraldTestnet = {
    id: 42261,
    name: "Emerald Paratime Testnet",
    network: "emerald-testnet",
    nativeCurrency: { name: "Emerald Rose", symbol: "ROSE", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://testnet.emerald.oasis.dev"] },
        public: { http: ["https://testnet.emerald.oasis.dev"] },
    },
    blockExplorers: {
        default: {
            name: "Oasis",
            url: "https://testnet.explorer.emerald.oasis.dev",
        },
    },
};
// https://docs.blast.io/building/network-information
exports.blast = {
    id: 81457,
    name: "Blast Mainnet",
    network: "blast-mainnet",
    nativeCurrency: { name: "Blast Ethereum", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://rpc.blast.io"] },
        public: { http: ["	https://rpc.blast.io"] },
    },
    blockExplorers: {
        default: {
            name: "Blast",
            url: "https://blastscan.io",
        },
    },
};
exports.blastSepolia = {
    id: 168587773,
    name: "Blast Sepolia",
    network: "blast-repolia",
    nativeCurrency: { name: "Blast Ethereum", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: { http: ["https://sepolia.blast.io"] },
        public: { http: ["https://sepolia.blast.io"] },
    },
    blockExplorers: {
        default: {
            name: "Blast Sepolia Explorer",
            url: "https://sepolia.blastexplorer.io",
        },
    },
};
exports.scroll = {
    id: 534352,
    name: "Scroll",
    network: "scroll",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.ankr.com/scroll"],
        },
        public: {
            http: ["https://rpc.ankr.com/scroll"],
        },
    },
    blockExplorers: {
        default: {
            name: "Scrollscan",
            url: "https://scrollscan.com",
        },
    },
};
exports.scrollSepolia = {
    id: 534351,
    name: "Scroll Sepolia",
    network: "scroll-sepolia",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.ankr.com/scroll_sepolia_testnet"],
        },
        public: {
            http: ["https://rpc.ankr.com/scroll_sepolia_testnet"],
        },
    },
    blockExplorers: {
        default: {
            name: "Scrollscan",
            url: "https://sepolia.scrollscan.com",
        },
    },
    testnet: true,
};
exports.xlayer = {
    id: 196,
    name: "X Layer",
    network: "xlayer-mainnet",
    nativeCurrency: { name: "OKB", symbol: "OKB", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.xlayer.tech"],
        },
        public: {
            http: ["https://rpc.xlayer.tech"],
        },
    },
    blockExplorers: {
        default: {
            name: "X Layer Explorer",
            url: "https://www.okx.com/web3/explorer/xlayer",
        },
    },
};
exports.xlayerTestnet = {
    id: 195,
    name: "X Layer Testnet",
    network: "xlayer-testnet",
    nativeCurrency: { name: "OKB", symbol: "OKB", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://testrpc.xlayer.tech"],
        },
        public: {
            http: ["https://testrpc.xlayer.tech"],
        },
    },
    blockExplorers: {
        default: {
            name: "X Layer Testnet Explorer",
            url: "https://www.okx.com/web3/explorer/xlayer-test",
        },
    },
    testnet: true,
};
exports.mantle = {
    id: 5000,
    name: "Mantle",
    network: "mantle",
    nativeCurrency: { name: "MNT", symbol: "MNT", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.mantle.xyz"],
        },
        public: {
            http: ["https://rpc.mantle.xyz"],
        },
    },
    blockExplorers: {
        default: {
            name: "Mantle Explorer",
            url: "https://explorer.mantle.xyz",
        },
    },
};
exports.mantleTestnet = {
    id: 5000,
    name: "Mantle",
    network: "mantle-testnet",
    nativeCurrency: { name: "MNT", symbol: "MNT", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.testnet.mantle.xyz"],
        },
        public: {
            http: ["https://rpc.testnet.mantle.xyz"],
        },
    },
    blockExplorers: {
        default: {
            name: "Mantle Explorer",
            url: "https://explorer.testnet.mantle.xyz",
        },
    },
};
exports.worldchain = {
    id: 480,
    name: "World Chain",
    network: "worldchain",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://worldchain-mainnet.g.alchemy.com/public"],
        },
        public: {
            http: ["https://worldchain-mainnet.g.alchemy.com/public"],
        },
    },
    blockExplorers: {
        default: {
            name: "World Scan",
            url: "https://worldscan.org/",
        },
    },
};
exports.worldchainTestnet = {
    id: 4801,
    name: "World Chain",
    network: "worldchain-sepolia",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://worldchain-sepolia.g.alchemy.com/public"],
        },
        public: {
            http: ["https://worldchain-sepolia.g.alchemy.com/public"],
        },
    },
    blockExplorers: {
        default: {
            name: "World Scan",
            url: "https://worldchain-sepolia.explorer.alchemy.com/",
        },
    },
    testnet: true,
};
exports.unichain = {
    id: 130,
    name: "Unichain",
    network: "unichain",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://mainnet.unichain.org"],
        },
        public: {
            http: ["https://mainnet.unichain.org"],
        },
    },
    blockExplorers: {
        default: {
            name: "Unichain Explorer",
            url: "https://unichain.blockscout.com/",
        },
    },
};
exports.unichainTestnet = {
    id: 1301,
    name: "Unichain Sepolia",
    network: "unichain-sepolia",
    nativeCurrency: { name: "Ether", symbol: "ETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://sepolia.unichain.org"],
        },
        public: {
            http: ["https://sepolia.unichain.org"],
        },
    },
    blockExplorers: {
        default: {
            name: "Unichain Explorer",
            url: "https://unichain-sepolia.blockscout.com/",
        },
    },
    testnet: true,
};
exports.monadTestnet = {
    id: 10143,
    name: "Monad Testnet",
    network: "monad-testnet",
    nativeCurrency: { name: "MON", symbol: "MON", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://testnet-rpc.monad.xyz"],
        },
        public: {
            http: ["https://testnet-rpc.monad.xyz"],
        },
    },
    blockExplorers: {
        default: {
            name: "Oasis",
            url: "https://testnet.explorer.oasis.dev",
        },
    },
    testnet: true,
};
exports.DEFAULT_CHAINS = [
    ...Object.values(CHAINS),
    acala,
    acalaTestnet,
    exports.blast,
    exports.blastSepolia,
    exports.emerald,
    exports.emeraldTestnet,
    karura,
    karuraTestnet,
    exports.scroll,
    exports.scrollSepolia,
    exports.xlayer,
    exports.xlayerTestnet,
    exports.mantle,
    exports.mantleTestnet,
    exports.worldchain,
    exports.worldchainTestnet,
    exports.unichain,
    exports.unichainTestnet,
    exports.monadTestnet,
];
