// @wormhole-foundation/wormhole-connect/lib/index.mjs
import { di as d, dk as s, dg as t, dj as T, dc as E, dA as e, dB as o, dx as _, du as A, dz as u, dv as N, dw as C, dD as n, dE as O, dy as U, dC as S, d8 as I, dl as i, dn as M, dq as r, dp as D, dm as P, d9 as c, db as l, dh as m, da as L, d7 as g, ds as y, dt as F, de as f, df as k, dd as B, dr as H } from "./index-CGIqRhuR.mjs";
export {
  d as AutomaticCCTPRoute,
  s as AutomaticPorticoRoute,
  t as AutomaticTokenBridgeRoute,
  T as CCTPRoute,
  E as DEFAULT_ROUTES,
  e as ERR_AMOUNT_TOO_LARGE,
  o as ERR_AMOUNT_TOO_SMALL,
  _ as ERR_DESTINATION_CONTRACT_PAUSED,
  A as ERR_INSUFFICIENT_ALLOWANCE,
  u as ERR_INSUFFICIENT_GAS,
  N as ERR_NOT_ENOUGH_CAPACITY,
  C as ERR_SOURCE_CONTRACT_PAUSED,
  n as ERR_TIMEOUT,
  O as ERR_UNKNOWN,
  U as ERR_UNSUPPORTED_ABI_VERSION,
  S as ERR_USER_REJECTED,
  I as MAINNET,
  i as MayanRoute,
  M as MayanRouteMCTP,
  r as MayanRouteSHUTTLE,
  D as MayanRouteSWIFT,
  P as MayanRouteWH,
  c as TESTNET,
  l as Token,
  m as TokenBridgeRoute,
  L as buildConfig,
  g as default,
  y as getTokenDetails,
  F as getTransferDetails,
  f as nttAutomaticRoute,
  k as nttManualRoute,
  B as nttRoutes,
  H as wormholeConnectHosted
};
