// @wormhole-foundation/wormhole-connect/lib/index-BjXRa_Tk.mjs
var _t = Object.defineProperty;
var Pt = (e, t, r) => t in e ? _t(e, t, { enumerable: !0, configurable: !0, writable: !0, value: r }) : e[t] = r;
var f = (e, t, r) => Pt(e, typeof t != "symbol" ? t + "" : t, r);
import { a9 as c, O as l, aW as A, bF as le, a_ as z, aZ as U, a$ as Z, aY as S, aX as V, bX as Ze, bY as Ge, bZ as Wt, b_ as It, b$ as Ut, c0 as Nt, bR as O, bE as E, b2 as y, c1 as T, c2 as Ft, c3 as ye, bx as dt, c4 as xt, bG as lt, bH as zt, c5 as Ot, c6 as X, c7 as Lt, bV as G, bW as H, c8 as Dt, bS as qt, c9 as Kt, ca as Vt, cb as Se, cc as Zt, bJ as v, bU as Gt, e as oe, f as k, bw as Ce, b9 as Ht, m as $, bC as Be, cd as Yt, ce as Xt, ao as ae, ad as L, cf as Qt, bK as ft, bf as mt, g as gt } from "./index-CGIqRhuR.mjs";
const M = new c("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
new c("TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb");
const Ae = new c("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"), D = new c("So11111111111111111111111111111111111111112");
new c("9pan9bMn5HatX4EJdBwg9VgCa7Uz5HL8N1m5D3NdXejP");
class ee extends Error {
  constructor(t) {
    super(t);
  }
}
class Re extends ee {
  constructor() {
    super(...arguments), this.name = "TokenAccountNotFoundError";
  }
}
class Jt extends ee {
  constructor() {
    super(...arguments), this.name = "TokenInvalidAccountError";
  }
}
class ht extends ee {
  constructor() {
    super(...arguments), this.name = "TokenInvalidAccountOwnerError";
  }
}
class ce extends ee {
  constructor() {
    super(...arguments), this.name = "TokenInvalidAccountSizeError";
  }
}
class $t extends ee {
  constructor() {
    super(...arguments), this.name = "TokenInvalidMintError";
  }
}
class pt extends ee {
  constructor() {
    super(...arguments), this.name = "TokenOwnerOffCurveError";
  }
}
var j;
(function(e) {
  e[e.InitializeMint = 0] = "InitializeMint", e[e.InitializeAccount = 1] = "InitializeAccount", e[e.InitializeMultisig = 2] = "InitializeMultisig", e[e.Transfer = 3] = "Transfer", e[e.Approve = 4] = "Approve", e[e.Revoke = 5] = "Revoke", e[e.SetAuthority = 6] = "SetAuthority", e[e.MintTo = 7] = "MintTo", e[e.Burn = 8] = "Burn", e[e.CloseAccount = 9] = "CloseAccount", e[e.FreezeAccount = 10] = "FreezeAccount", e[e.ThawAccount = 11] = "ThawAccount", e[e.TransferChecked = 12] = "TransferChecked", e[e.ApproveChecked = 13] = "ApproveChecked", e[e.MintToChecked = 14] = "MintToChecked", e[e.BurnChecked = 15] = "BurnChecked", e[e.InitializeAccount2 = 16] = "InitializeAccount2", e[e.SyncNative = 17] = "SyncNative", e[e.InitializeAccount3 = 18] = "InitializeAccount3", e[e.InitializeMultisig2 = 19] = "InitializeMultisig2", e[e.InitializeMint2 = 20] = "InitializeMint2", e[e.GetAccountDataSize = 21] = "GetAccountDataSize", e[e.InitializeImmutableOwner = 22] = "InitializeImmutableOwner", e[e.AmountToUiAmount = 23] = "AmountToUiAmount", e[e.UiAmountToAmount = 24] = "UiAmountToAmount", e[e.InitializeMintCloseAuthority = 25] = "InitializeMintCloseAuthority", e[e.TransferFeeExtension = 26] = "TransferFeeExtension", e[e.ConfidentialTransferExtension = 27] = "ConfidentialTransferExtension", e[e.DefaultAccountStateExtension = 28] = "DefaultAccountStateExtension", e[e.Reallocate = 29] = "Reallocate", e[e.MemoTransferExtension = 30] = "MemoTransferExtension", e[e.CreateNativeMint = 31] = "CreateNativeMint", e[e.InitializeNonTransferableMint = 32] = "InitializeNonTransferableMint", e[e.InterestBearingMintExtension = 33] = "InterestBearingMintExtension", e[e.CpiGuardExtension = 34] = "CpiGuardExtension", e[e.InitializePermanentDelegate = 35] = "InitializePermanentDelegate", e[e.TransferHookExtension = 36] = "TransferHookExtension", e[e.MetadataPointerExtension = 39] = "MetadataPointerExtension";
})(j || (j = {}));
function Ee(e, t, r) {
  if (r.length) {
    e.push({ pubkey: t, isSigner: !1, isWritable: !1 });
    for (const n of r)
      e.push({
        pubkey: n instanceof c ? n : n.publicKey,
        isSigner: !0,
        isWritable: !1
      });
  } else
    e.push({ pubkey: t, isSigner: !0, isWritable: !1 });
  return e;
}
const He = A([z("instruction"), U("amount")]);
function jt(e, t, r, n, i = [], a = M) {
  const o = Ee([
    { pubkey: e, isSigner: !1, isWritable: !0 },
    { pubkey: t, isSigner: !1, isWritable: !1 }
  ], r, i), s = l.Buffer.alloc(He.span);
  return He.encode({
    instruction: j.Approve,
    amount: BigInt(n)
  }, s), new le({ keys: o, programId: a, data: s });
}
const Ye = A([z("instruction")]);
function Xe(e, t, r, n = [], i = M) {
  const a = Ee([
    { pubkey: e, isSigner: !1, isWritable: !0 },
    { pubkey: t, isSigner: !1, isWritable: !0 }
  ], r, n), o = l.Buffer.alloc(Ye.span);
  return Ye.encode({ instruction: j.CloseAccount }, o), new le({ keys: a, programId: i, data: o });
}
var ge;
(function(e) {
  e[e.Uninitialized = 0] = "Uninitialized", e[e.Mint = 1] = "Mint", e[e.Account = 2] = "Account";
})(ge || (ge = {}));
const _e = 1, en = A([
  z("m"),
  z("n"),
  Z("isInitialized"),
  S("signer1"),
  S("signer2"),
  S("signer3"),
  S("signer4"),
  S("signer5"),
  S("signer6"),
  S("signer7"),
  S("signer8"),
  S("signer9"),
  S("signer10"),
  S("signer11")
]), Pe = en.span;
var he;
(function(e) {
  e[e.Uninitialized = 0] = "Uninitialized", e[e.Initialized = 1] = "Initialized", e[e.Frozen = 2] = "Frozen";
})(he || (he = {}));
const wt = A([
  S("mint"),
  S("owner"),
  U("amount"),
  V("delegateOption"),
  S("delegate"),
  z("state"),
  V("isNativeOption"),
  U("isNative"),
  U("delegatedAmount"),
  V("closeAuthorityOption"),
  S("closeAuthority")
]), P = wt.span;
async function tn(e, t, r, n = M) {
  const i = await e.getAccountInfo(t, r);
  return rn(t, i, n);
}
async function Qe(e, t) {
  return await nn(e, [], t);
}
async function nn(e, t, r) {
  const n = zn(t);
  return await e.getMinimumBalanceForRentExemption(n, r);
}
function rn(e, t, r = M) {
  if (!t)
    throw new Re();
  if (!t.owner.equals(r))
    throw new ht();
  if (t.data.length < P)
    throw new ce();
  const n = wt.decode(t.data.slice(0, P));
  let i = l.Buffer.alloc(0);
  if (t.data.length > P) {
    if (t.data.length === Pe)
      throw new ce();
    if (t.data[P] != ge.Account)
      throw new Jt();
    i = t.data.slice(P + _e);
  }
  return {
    address: e,
    mint: n.mint,
    owner: n.owner,
    amount: n.amount,
    delegate: n.delegateOption ? n.delegate : null,
    delegatedAmount: n.delegatedAmount,
    isInitialized: n.state !== he.Uninitialized,
    isFrozen: n.state === he.Frozen,
    isNative: !!n.isNativeOption,
    rentExemptReserve: n.isNativeOption ? n.isNative : null,
    closeAuthority: n.closeAuthorityOption ? n.closeAuthority : null,
    tlvData: i
  };
}
const yt = A([
  V("mintAuthorityOption"),
  S("mintAuthority"),
  U("supply"),
  z("decimals"),
  Z("isInitialized"),
  V("freezeAuthorityOption"),
  S("freezeAuthority")
]), ve = yt.span;
async function an(e, t, r, n = M) {
  const i = await e.getAccountInfo(t, r);
  return sn(t, i, n);
}
function sn(e, t, r = M) {
  if (!t)
    throw new Re();
  if (!t.owner.equals(r))
    throw new ht();
  if (t.data.length < ve)
    throw new ce();
  const n = yt.decode(t.data.slice(0, ve));
  let i = l.Buffer.alloc(0);
  if (t.data.length > ve) {
    if (t.data.length <= P)
      throw new ce();
    if (t.data.length === Pe)
      throw new ce();
    if (t.data[P] != ge.Mint)
      throw new $t();
    i = t.data.slice(P + _e);
  }
  return {
    address: e,
    mintAuthority: n.mintAuthorityOption ? n.mintAuthority : null,
    supply: n.supply,
    decimals: n.decimals,
    isInitialized: n.isInitialized,
    freezeAuthority: n.freezeAuthorityOption ? n.freezeAuthority : null,
    tlvData: i
  };
}
async function Je(e, t, r = !1, n = M, i = Ae) {
  if (!r && !c.isOnCurve(t.toBuffer()))
    throw new pt();
  const [a] = await c.findProgramAddress([t.toBuffer(), n.toBuffer(), e.toBuffer()], i);
  return a;
}
function We(e, t, r = !1, n = M, i = Ae) {
  if (!r && !c.isOnCurve(t.toBuffer()))
    throw new pt();
  const [a] = c.findProgramAddressSync([t.toBuffer(), n.toBuffer(), e.toBuffer()], i);
  return a;
}
const on = A([Z("lockCpi")]), cn = on.span, un = A([z("state")]), dn = un.span, ln = A([]), fn = ln.span, mn = A([
  S("rateAuthority"),
  Ze("initializationTimestamp"),
  Ge("preUpdateAverageRate"),
  Ze("lastUpdateTimestamp"),
  Ge("currentRate")
]), gn = mn.span, hn = A([Z("requireIncomingTransferMemos")]), pn = hn.span, wn = A([
  S("authority"),
  S("metadataAddress")
]), yn = wn.span, Sn = A([S("closeAuthority")]), An = Sn.span, St = A([]), Mn = St.span, kn = St.span, bn = A([S("delegate")]), Tn = bn.span, Cn = 1e4;
BigInt(Cn);
function $e(e) {
  return A([U("epoch"), U("maximumFee"), Wt("transferFeeBasisPoints")], e);
}
const vn = A([
  S("transferFeeConfigAuthority"),
  S("withdrawWithheldAuthority"),
  U("withheldAmount"),
  $e("olderTransferFee"),
  $e("newerTransferFee")
]), Bn = vn.span, Rn = A([U("withheldAmount")]), En = Rn.span, _n = A([S("authority"), S("programId")]), Pn = _n.span, Wn = A([Z("transferring")]), In = Wn.span, je = A([
  z("discriminator"),
  It(32, "addressConfig"),
  Z("isSigner"),
  Z("isWritable")
]), Un = A([
  V("count"),
  Ut(je, Nt(je.span), "extraAccounts")
]);
A([
  U("instructionDiscriminator"),
  V("length"),
  Un.replicate("extraAccountsList")
]);
var b;
(function(e) {
  e[e.Uninitialized = 0] = "Uninitialized", e[e.TransferFeeConfig = 1] = "TransferFeeConfig", e[e.TransferFeeAmount = 2] = "TransferFeeAmount", e[e.MintCloseAuthority = 3] = "MintCloseAuthority", e[e.ConfidentialTransferMint = 4] = "ConfidentialTransferMint", e[e.ConfidentialTransferAccount = 5] = "ConfidentialTransferAccount", e[e.DefaultAccountState = 6] = "DefaultAccountState", e[e.ImmutableOwner = 7] = "ImmutableOwner", e[e.MemoTransfer = 8] = "MemoTransfer", e[e.NonTransferable = 9] = "NonTransferable", e[e.InterestBearingConfig = 10] = "InterestBearingConfig", e[e.CpiGuard = 11] = "CpiGuard", e[e.PermanentDelegate = 12] = "PermanentDelegate", e[e.NonTransferableAccount = 13] = "NonTransferableAccount", e[e.TransferHook = 14] = "TransferHook", e[e.TransferHookAccount = 15] = "TransferHookAccount", e[e.MetadataPointer = 18] = "MetadataPointer";
})(b || (b = {}));
const et = 2, Nn = 2;
function Fn(e) {
  switch (e) {
    case b.Uninitialized:
      return 0;
    case b.TransferFeeConfig:
      return Bn;
    case b.TransferFeeAmount:
      return En;
    case b.MintCloseAuthority:
      return An;
    case b.ConfidentialTransferMint:
      return 97;
    case b.ConfidentialTransferAccount:
      return 286;
    case b.CpiGuard:
      return cn;
    case b.DefaultAccountState:
      return dn;
    case b.ImmutableOwner:
      return fn;
    case b.MemoTransfer:
      return pn;
    case b.MetadataPointer:
      return yn;
    case b.NonTransferable:
      return Mn;
    case b.InterestBearingConfig:
      return gn;
    case b.PermanentDelegate:
      return Tn;
    case b.NonTransferableAccount:
      return kn;
    case b.TransferHook:
      return Pn;
    case b.TransferHookAccount:
      return In;
    default:
      throw Error(`Unknown extension type: ${e}`);
  }
}
function xn(e, t) {
  if (e.length === 0)
    return t;
  {
    const r = P + _e + e.filter((n, i) => i === e.indexOf(n)).map((n) => Fn(n) + et + Nn).reduce((n, i) => n + i);
    return r === Pe ? r + et : r;
  }
}
function zn(e) {
  return xn(e, P);
}
const tt = A([z("instruction")]);
function nt(e, t, r, n = M) {
  const i = [
    { pubkey: e, isSigner: !1, isWritable: !0 },
    { pubkey: t, isSigner: !1, isWritable: !1 },
    { pubkey: r, isSigner: !1, isWritable: !1 },
    { pubkey: O, isSigner: !1, isWritable: !1 }
  ], a = l.Buffer.alloc(tt.span);
  return tt.encode({ instruction: j.InitializeAccount }, a), new le({ keys: i, programId: n, data: a });
}
function At(e, t, r, n, i = M, a = Ae) {
  return On(e, t, r, n, l.Buffer.alloc(0), i, a);
}
function On(e, t, r, n, i, a = M, o = Ae) {
  const s = [
    { pubkey: e, isSigner: !0, isWritable: !0 },
    { pubkey: t, isSigner: !1, isWritable: !0 },
    { pubkey: r, isSigner: !1, isWritable: !1 },
    { pubkey: n, isSigner: !1, isWritable: !1 },
    { pubkey: E.programId, isSigner: !1, isWritable: !1 },
    { pubkey: a, isSigner: !1, isWritable: !1 }
  ];
  return new le({
    keys: s,
    programId: o,
    data: i
  });
}
const rt = A([z("instruction"), U("amount")]);
function Ln(e, t, r, n, i = [], a = M) {
  const o = Ee([
    { pubkey: e, isSigner: !1, isWritable: !0 },
    { pubkey: t, isSigner: !1, isWritable: !0 }
  ], r, i), s = l.Buffer.alloc(rt.span);
  return rt.encode({
    instruction: j.Transfer,
    amount: BigInt(n)
  }, s), new le({ keys: o, programId: a, data: s });
}
function Ie(e, t) {
  const r = l.Buffer.alloc(2);
  return r.writeUInt16BE(y(t)), T([l.Buffer.from("foreign_contract"), r], e);
}
function Dn(e) {
  return T([l.Buffer.from("redeemer")], e);
}
function Ue(e, t) {
  return T([l.Buffer.from("mint"), new c(t).toBuffer()], e);
}
function Mt(e) {
  return T([l.Buffer.from("sender")], e);
}
function kt(e, t) {
  return T([l.Buffer.from("tmp"), new c(t).toBuffer()], e);
}
function bt(e, t, r) {
  return T([
    l.Buffer.from("bridged"),
    new c(t).toBuffer(),
    l.Buffer.from(Ft.toBytes(BigInt(r.toString()), 8))
  ], e);
}
function B(e) {
  return T([l.Buffer.from("config")], e);
}
async function Rr(e, t, r) {
  return e.getAccountInfo(B(t), r).then((n) => Ne.deserialize(ye(n)));
}
class Ne {
  constructor(t) {
    f(this, "wormhole");
    this.wormhole = new c(t);
  }
  static deserialize(t) {
    if (t.length != 32)
      throw new Error("data.length != 32");
    const r = t.subarray(0, 32);
    return new Ne(r);
  }
}
function Me(e, t) {
  return T([new c(t).toBuffer()], e);
}
function te(e, t, r) {
  if (t == y("Solana"))
    throw new Error("emitterChain == CHAIN_ID_SOLANA cannot exist as foreign token bridge emitter");
  const n = typeof r == "string" ? new dt(r).toUint8Array() : r;
  return T([
    (() => {
      const i = l.Buffer.alloc(2);
      return i.writeUInt16BE(t), i;
    })(),
    n
  ], e);
}
async function Er(e, t, r) {
  return e.getAccountInfo(new c(t), r).then((n) => Fe.deserialize(ye(n)));
}
class Fe {
  constructor(t, r) {
    f(this, "chain");
    f(this, "contract");
    this.chain = t, this.contract = r;
  }
  static deserialize(t) {
    if (t.length != 34)
      throw new Error("data.length != 34");
    const r = t.readUInt16LE(0), n = t.subarray(2, 34);
    return new Fe(r, n);
  }
}
function xe(e) {
  return T([l.Buffer.from("sender")], e);
}
function ze(e) {
  return T([l.Buffer.from("redeemer")], e);
}
function ne(e) {
  return T([l.Buffer.from("authority_signer")], e);
}
function fe(e) {
  return T([l.Buffer.from("custody_signer")], e);
}
function ke(e) {
  return T([l.Buffer.from("mint_signer")], e);
}
function re(e, t, r) {
  if (t == y("Solana"))
    throw new Error("tokenChain == CHAIN_ID_SOLANA does not have wrapped mint key");
  return T([
    l.Buffer.from("wrapped"),
    (() => {
      const n = l.Buffer.alloc(2);
      return n.writeUInt16BE(t), n;
    })(),
    r
  ], e);
}
function Y(e, t) {
  return T([l.Buffer.from("meta"), new c(t).toBuffer()], e);
}
async function me(e, t, r, n) {
  return e.getAccountInfo(Y(t, r), n).then((i) => Oe.deserialize(ye(i)));
}
class Oe {
  constructor(t, r, n, i) {
    f(this, "chain");
    f(this, "tokenAddress");
    f(this, "originalDecimals");
    f(this, "lastUpdatedSequence");
    this.chain = t, this.tokenAddress = r, this.originalDecimals = n, this.lastUpdatedSequence = i;
  }
  static deserialize(t) {
    if (t.length !== 35 && t.length !== 43)
      throw new Error(`invalid wrapped meta length: ${t.length}`);
    const r = t.readUInt16LE(0), n = t.subarray(2, 34), i = t.readUInt8(34), a = t.length === 43 ? t.readBigUInt64LE(35) : void 0;
    return new Oe(r, n, i, a);
  }
}
function it(e, t, r, n) {
  return jt(new c(t), ne(e), new c(r), n);
}
class qn {
  constructor(t) {
    f(this, "idl");
    this.idl = t;
  }
  async encode(t, r) {
    switch (t) {
      default:
        throw new Error(`Invalid account name: ${t}`);
    }
  }
  decode(t, r) {
    return this.decodeUnchecked(t, r);
  }
  decodeUnchecked(t, r) {
    switch (t) {
      default:
        throw new Error(`Invalid account name: ${t}`);
    }
  }
  memcmp(t, r) {
    switch (t) {
      default:
        throw new Error(`Invalid account name: ${t}`);
    }
  }
  size(t) {
    return xt(this.idl, t) ?? 0;
  }
}
class Kn {
  constructor(t) {
  }
  decode(t) {
    throw new Error("Token Bridge program does not have events");
  }
}
class Vn {
  constructor(t) {
  }
  encode(t, r) {
    switch (t) {
      case "initialize":
        return Zn(r);
      case "attestToken":
        return Gn(r);
      case "completeNative":
        return Hn(r);
      case "completeWrapped":
        return Yn(r);
      case "transferWrapped":
        return Xn(r);
      case "transferNative":
        return Qn(r);
      case "registerChain":
        return Jn(r);
      case "createWrapped":
        return $n(r);
      case "upgradeContract":
        return jn(r);
      case "transferWrappedWithPayload":
        return er(r);
      case "transferNativeWithPayload":
        return tr(r);
      default:
        throw new Error(`Invalid instruction: ${t}`);
    }
  }
  encodeState(t, r) {
    throw new Error("Token Bridge program does not have state");
  }
}
var _;
(function(e) {
  e[e.Initialize = 0] = "Initialize", e[e.AttestToken = 1] = "AttestToken", e[e.CompleteNative = 2] = "CompleteNative", e[e.CompleteWrapped = 3] = "CompleteWrapped", e[e.TransferWrapped = 4] = "TransferWrapped", e[e.TransferNative = 5] = "TransferNative", e[e.RegisterChain = 6] = "RegisterChain", e[e.CreateWrapped = 7] = "CreateWrapped", e[e.UpgradeContract = 8] = "UpgradeContract", e[e.CompleteNativeWithPayload = 9] = "CompleteNativeWithPayload", e[e.CompleteWrappedWithPayload = 10] = "CompleteWrappedWithPayload", e[e.TransferWrappedWithPayload = 11] = "TransferWrappedWithPayload", e[e.TransferNativeWithPayload = 12] = "TransferNativeWithPayload";
})(_ || (_ = {}));
function N(e, t) {
  const r = t === void 0 ? 0 : t.length, n = l.Buffer.alloc(1 + r);
  return n.writeUInt8(e, 0), r > 0 && n.write(t.toString("hex"), 1, "hex"), n;
}
function Zn({ wormhole: e }) {
  const t = l.Buffer.alloc(32);
  return t.write(new c(e).toBuffer().toString("hex"), 0, "hex"), N(_.Initialize, t);
}
function Gn({ nonce: e }) {
  const t = l.Buffer.alloc(4);
  return t.writeUInt32LE(e, 0), N(_.AttestToken, t);
}
function Hn({}) {
  return N(_.CompleteNative);
}
function Yn({}) {
  return N(_.CompleteWrapped);
}
function Tt({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i }) {
  if (typeof t != "bigint" && (t = BigInt(t)), typeof r != "bigint" && (r = BigInt(r)), !l.Buffer.isBuffer(n))
    throw new Error("targetAddress must be Buffer");
  const a = l.Buffer.alloc(54);
  return a.writeUInt32LE(e, 0), a.writeBigUInt64LE(t, 4), a.writeBigUInt64LE(r, 12), a.write(n.toString("hex"), 20, "hex"), a.writeUInt16LE(i, 52), a;
}
function Xn({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i }) {
  return N(_.TransferWrapped, Tt({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i }));
}
function Qn({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i }) {
  return N(_.TransferNative, Tt({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i }));
}
function Jn({}) {
  return N(_.RegisterChain);
}
function $n({}) {
  return N(_.CreateWrapped);
}
function jn({}) {
  return N(_.UpgradeContract);
}
function Ct({ nonce: e, amount: t, targetAddress: r, targetChain: n, payload: i }) {
  if (typeof t != "bigint" && (t = BigInt(t)), !l.Buffer.isBuffer(r))
    throw new Error("targetAddress must be Buffer");
  if (!l.Buffer.isBuffer(i))
    throw new Error("payload must be Buffer");
  const a = l.Buffer.alloc(50);
  return a.writeUInt32LE(e, 0), a.writeBigUInt64LE(t, 4), a.write(r.toString("hex"), 12, "hex"), a.writeUInt16LE(n, 44), a.writeUInt32LE(i.length, 46), l.Buffer.concat([
    a,
    i,
    l.Buffer.alloc(1)
    // option == None
  ]);
}
function er({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i, payload: a }) {
  return N(_.TransferWrappedWithPayload, Ct({
    nonce: e,
    amount: t,
    targetAddress: n,
    targetChain: i,
    payload: a
  }));
}
function tr({ nonce: e, amount: t, fee: r, targetAddress: n, targetChain: i, payload: a }) {
  return N(_.TransferNativeWithPayload, Ct({
    nonce: e,
    amount: t,
    targetAddress: n,
    targetChain: i,
    payload: a
  }));
}
class nr {
  constructor(t) {
  }
  encode(t, r) {
    throw new Error("Token Bridge program does not have state");
  }
  decode(t) {
    throw new Error("Token Bridge program does not have state");
  }
}
class rr {
  constructor(t) {
  }
  encode(t, r) {
    throw new Error("Token Bridge program does not have user-defined types");
  }
  decode(t, r) {
    throw new Error("Token Bridge program does not have user-defined types");
  }
}
class ir {
  constructor(t) {
    f(this, "instruction");
    f(this, "accounts");
    f(this, "state");
    f(this, "events");
    f(this, "types");
    this.instruction = new Vn(t), this.accounts = new qn(t), this.state = new nr(t), this.events = new Kn(t), this.types = new rr(t);
  }
}
const vt = {
  version: "0.1.0",
  name: "wormhole",
  instructions: [
    {
      name: "initialize",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "wormhole",
          type: "publicKey"
        }
      ]
    },
    {
      name: "attestToken",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wrappedMeta",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "splMetadata",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "wormholeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "nonce",
          type: "u32"
        }
      ]
    },
    {
      name: "completeNative",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "claim",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "endpoint",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "to",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "toFees",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "custody",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "custodySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: []
    },
    {
      name: "completeWrapped",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "claim",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "endpoint",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "to",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "toFees",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wrappedMeta",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "mintAuthority",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: []
    },
    {
      name: "transferWrapped",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "from",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "fromOwner",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wrappedMeta",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "authoritySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "wormholeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "nonce",
          type: "u32"
        },
        {
          name: "amount",
          type: "u64"
        },
        {
          name: "fee",
          type: "u64"
        },
        {
          name: "targetAddress",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "targetChain",
          type: "u16"
        }
      ]
    },
    {
      name: "transferNative",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "from",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "custody",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "authoritySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "custodySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "wormholeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "nonce",
          type: "u32"
        },
        {
          name: "amount",
          type: "u64"
        },
        {
          name: "fee",
          type: "u64"
        },
        {
          name: "targetAddress",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "targetChain",
          type: "u16"
        }
      ]
    },
    {
      name: "registerChain",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "endpoint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "claim",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: []
    },
    {
      name: "createWrapped",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "endpoint",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "claim",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wrappedMeta",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "splMetadata",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mintAuthority",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "splMetadataProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: []
    },
    {
      name: "upgradeContract",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "claim",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "upgradeAuthority",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "spill",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "implementation",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "programData",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "tokenBridgeProgram",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "bpfLoaderUpgradeable",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: []
    },
    {
      name: "transferWrappedWithPayload",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "from",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "fromOwner",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wrappedMeta",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "authoritySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "wormholeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "sender",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "nonce",
          type: "u32"
        },
        {
          name: "amount",
          type: "u64"
        },
        {
          name: "targetAddress",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "targetChain",
          type: "u16"
        },
        {
          name: "payload",
          type: "bytes"
        },
        {
          name: "cpiProgramId",
          type: {
            option: "publicKey"
          }
        }
      ]
    },
    {
      name: "transferNativeWithPayload",
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "from",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "custody",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "authoritySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "custodySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !0
        },
        {
          name: "wormholeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "sender",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "nonce",
          type: "u32"
        },
        {
          name: "amount",
          type: "u64"
        },
        {
          name: "targetAddress",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "targetChain",
          type: "u16"
        },
        {
          name: "payload",
          type: "bytes"
        },
        {
          name: "cpiProgramId",
          type: {
            option: "publicKey"
          }
        }
      ]
    }
  ],
  accounts: []
};
function ar(e, t) {
  return new lt(vt, new c(e), t === void 0 ? { connection: null } : t, sr());
}
function W(e, t) {
  return ar(e, zt(t));
}
function sr() {
  return new ir(vt);
}
const ue = class ue {
  constructor(t, r, n) {
    f(this, "address");
    f(this, "verified");
    f(this, "share");
    this.address = new c(t), this.verified = r, this.share = n;
  }
  serialize() {
    const t = l.Buffer.alloc(ue.size);
    return t.write(this.address.toBuffer().toString("hex"), 0, "hex"), this.verified && t.writeUInt8(1, 32), t.writeUInt8(this.share, 33), t;
  }
  static deserialize(t) {
    const r = t.subarray(0, 32), n = t.readUInt8(32) > 0, i = t.readUInt8(33);
    return new ue(r, n, i);
  }
};
f(ue, "size", 34);
let K = ue;
class be {
  constructor(t, r, n, i, a) {
    f(this, "name");
    f(this, "symbol");
    f(this, "uri");
    f(this, "sellerFeeBasisPoints");
    f(this, "creators");
    this.name = t, this.symbol = r, this.uri = n, this.sellerFeeBasisPoints = i, this.creators = a;
  }
  serialize() {
    const t = this.name.length, r = this.symbol.length, n = this.uri.length, i = this.creators, [a, o] = (() => {
      if (i === null)
        return [0, 0];
      const d = i.length;
      return [d, 4 + d * K.size];
    })(), s = l.Buffer.alloc(15 + t + r + n + o);
    if (s.writeUInt32LE(t, 0), s.write(this.name, 4), s.writeUInt32LE(r, 4 + t), s.write(this.symbol, 8 + t), s.writeUInt32LE(n, 8 + t + r), s.write(this.uri, 12 + t + r), s.writeUInt16LE(this.sellerFeeBasisPoints, 12 + t + r + n), i === null)
      s.writeUInt8(0, 14 + t + r + n);
    else {
      s.writeUInt8(1, 14 + t + r + n), s.writeUInt32LE(a, 15 + t + r + n);
      for (let d = 0; d < a; ++d) {
        const u = i.at(d), m = 19 + t + r + n + d * K.size;
        s.write(u.serialize().toString("hex"), m, "hex");
      }
    }
    return s;
  }
  static deserialize(t) {
    const r = t.readUInt32LE(0), n = t.subarray(4, 4 + r).toString(), i = t.readUInt32LE(4 + r), a = t.subarray(8 + r, 8 + r + i).toString(), o = t.readUInt32LE(8 + r + i), s = t.subarray(12 + r + i, 12 + r + i + o).toString(), d = t.readUInt16LE(12 + r + i + o), u = t.readUInt8(14 + r + i + o), m = (() => {
      if (u == 0)
        return null;
      const h = [], g = t.readUInt32LE(15 + r + i + o);
      for (let w = 0; w < g; ++w) {
        const p = 19 + r + i + o + w * K.size;
        h.push(K.deserialize(t.subarray(p, p + K.size)));
      }
      return h;
    })();
    return new be(n, a, s, d, m);
  }
}
class pe extends be {
  constructor(r, n, i, a, o, s) {
    super(r, n, i, a, o);
    f(this, "isMutable");
    this.isMutable = s;
  }
  static serialize(r, n, i, a, o, s) {
    return new pe(r, n, i, a, o, s).serialize();
  }
  static serializeInstructionData(r, n, i, a, o, s) {
    return l.Buffer.concat([
      l.Buffer.alloc(1, 0),
      pe.serialize(r, n, i, a, o, s)
    ]);
  }
  serialize() {
    return l.Buffer.concat([
      super.serialize(),
      l.Buffer.alloc(1, this.isMutable ? 1 : 0)
    ]);
  }
}
const we = class we {
  /**
   * @internal
   */
  constructor() {
  }
  static createMetadataAccounts(t, r, n, i, a, o, s = !1, d, u, m, h = !1, g = Te(r)) {
    const w = [
      Ot(g, !1),
      X(r, !1),
      X(n, !0),
      X(t, !0),
      X(o, s),
      X(E.programId, !1),
      X(O, !1)
    ], p = pe.serializeInstructionData(i, a, d === void 0 ? "" : d, m === void 0 ? 0 : m, u === void 0 ? null : u, h);
    return {
      programId: we.programId,
      keys: w,
      data: p
    };
  }
};
/**
 * Public key that identifies the SPL Token Metadata program
 */
f(we, "programId", new c("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s"));
let de = we;
function Te(e) {
  return T([
    l.Buffer.from("metadata"),
    de.programId.toBuffer(),
    new c(e).toBuffer()
  ], de.programId);
}
var at;
(function(e) {
  e[e.Uninitialized = 0] = "Uninitialized", e[e.EditionV1 = 1] = "EditionV1", e[e.MasterEditionV1 = 2] = "MasterEditionV1", e[e.ReservationListV1 = 3] = "ReservationListV1", e[e.MetadataV1 = 4] = "MetadataV1", e[e.ReservationListV2 = 5] = "ReservationListV2", e[e.MasterEditionV2 = 6] = "MasterEditionV2", e[e.EditionMarker = 7] = "EditionMarker";
})(at || (at = {}));
class Le {
  constructor(t, r, n, i, a, o) {
    f(this, "key");
    f(this, "updateAuthority");
    f(this, "mint");
    f(this, "data");
    f(this, "primarySaleHappened");
    f(this, "isMutable");
    this.key = t, this.updateAuthority = new c(r), this.mint = new c(n), this.data = i, this.primarySaleHappened = a, this.isMutable = o;
  }
  static deserialize(t) {
    const r = t.readUInt8(0), n = t.subarray(1, 33), i = t.subarray(33, 65), a = be.deserialize(t.subarray(65)), o = a.serialize().length, s = t.readUInt8(65 + o) > 0, d = t.readUInt8(66 + o) > 0;
    return new Le(r, n, i, a, s, d);
  }
}
async function _r(e, t, r) {
  return e.getAccountInfo(Te(t), r).then((n) => Le.deserialize(ye(n)));
}
function or(e, t, r, n, i, a, o) {
  const s = W(t, e).methods.attestToken(o);
  return console.log(st(t, r, n, i, a)), s._ixFn(...s._args, {
    accounts: st(t, r, n, i, a),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function st(e, t, r, n, i) {
  const { bridge: a, emitter: o, sequence: s, feeCollector: d, clock: u, rent: m, systemProgram: h } = Lt(t, r, i, e);
  return {
    payer: new c(r),
    config: B(e),
    mint: new c(n),
    wrappedMeta: Y(e, n),
    splMetadata: Te(n),
    wormholeBridge: a,
    wormholeMessage: new c(i),
    wormholeEmitter: o,
    wormholeSequence: s,
    wormholeFeeCollector: d,
    clock: u,
    rent: m,
    systemProgram: h,
    wormholeProgram: new c(t)
  };
}
function ot(e, t, r, n, i, a) {
  const o = W(t, e).methods.completeNative();
  return o._ixFn(...o._args, {
    accounts: cr(t, r, n, i, a),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function cr(e, t, r, n, i) {
  const a = new c(n.payload.token.address.toUint8Array());
  return {
    payer: new c(r),
    config: B(e),
    vaa: H(t, l.Buffer.from(n.hash)),
    claim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    endpoint: te(e, y(n.emitterChain), n.emitterAddress.toUint8Array()),
    to: new c(n.payload.to.address.toUint8Array()),
    toFees: new c(i === void 0 ? n.payload.to.address.toUint8Array() : i),
    custody: Me(e, a),
    mint: a,
    custodySigner: fe(e),
    rent: O,
    systemProgram: E.programId,
    tokenProgram: M,
    wormholeProgram: new c(t)
  };
}
function ur(e, t, r, n, i, a) {
  const o = W(t, e).methods.completeWrapped();
  return o._ixFn(...o._args, {
    accounts: dr(t, r, n, i, a),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function dr(e, t, r, n, i) {
  const a = re(e, y(n.payload.token.chain), n.payload.token.address.toUint8Array());
  return {
    payer: new c(r),
    config: B(e),
    vaa: H(t, l.Buffer.from(n.hash)),
    claim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    endpoint: te(e, y(n.emitterChain), n.emitterAddress.toUint8Array()),
    to: new c(n.payload.to.address.toUint8Array()),
    toFees: new c(i === void 0 ? n.payload.to.address.toUint8Array() : i),
    mint: a,
    wrappedMeta: Y(e, a),
    mintAuthority: ke(e),
    rent: O,
    systemProgram: E.programId,
    tokenProgram: M,
    wormholeProgram: new c(t)
  };
}
function lr(e, t, r, n, i) {
  const a = W(t, e).methods.createWrapped();
  return a._ixFn(...a._args, {
    accounts: fr(t, r, n, i),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function fr(e, t, r, n) {
  const i = re(e, y(n.payload.token.chain), n.payload.token.address.toUint8Array());
  return {
    payer: new c(r),
    config: B(e),
    endpoint: te(e, y(n.emitterChain), n.emitterAddress.toUint8Array()),
    vaa: H(t, l.Buffer.from(n.hash)),
    claim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    mint: i,
    wrappedMeta: Y(e, i),
    splMetadata: Te(i),
    mintAuthority: ke(e),
    rent: O,
    systemProgram: E.programId,
    tokenProgram: M,
    splMetadataProgram: de.programId,
    wormholeProgram: new c(t)
  };
}
function Pr(e, t, r) {
  const n = W(e).methods.initialize(r);
  return n._ixFn(...n._args, {
    accounts: mr(e, t),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function mr(e, t) {
  return {
    payer: new c(t),
    config: B(e),
    rent: O,
    systemProgram: E.programId
  };
}
function Wr(e, t, r, n) {
  const i = W(e).methods.registerChain();
  return i._ixFn(...i._args, {
    accounts: gr(e, t, r, n),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function gr(e, t, r, n) {
  return {
    payer: new c(r),
    config: B(e),
    endpoint: te(e, y(n.payload.actionArgs.foreignChain), n.payload.actionArgs.foreignAddress.toUint8Array()),
    vaa: H(t, l.Buffer.from(n.hash)),
    claim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    rent: O,
    systemProgram: E.programId,
    wormholeProgram: new c(t)
  };
}
function Ir(e, t, r, n, i) {
  const a = W(e).methods.upgradeContract();
  return a._ixFn(...a._args, {
    accounts: hr(e, t, r, n, i),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function hr(e, t, r, n, i) {
  return {
    payer: new c(r),
    vaa: H(t, l.Buffer.from(n.hash)),
    claim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    upgradeAuthority: Vt(e),
    spill: new c(i === void 0 ? r : i),
    implementation: new c(n.payload.actionArgs.newContract),
    programData: Kt(e),
    tokenBridgeProgram: new c(e),
    rent: O,
    clock: qt,
    bpfLoaderUpgradeable: Dt,
    systemProgram: E.programId
  };
}
function ct(e, t, r, n, i, a, o, s, d, u, m, h) {
  const g = W(t, e).methods.transferNative(s, d, u, l.Buffer.from(m), h);
  return g._ixFn(...g._args, {
    accounts: pr(t, r, n, i, a, o),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function pr(e, t, r, n, i, a) {
  const { wormholeBridge: o, wormholeMessage: s, wormholeEmitter: d, wormholeSequence: u, wormholeFeeCollector: m, clock: h, rent: g, systemProgram: w } = Se(e, t, r, n);
  return {
    payer: new c(r),
    config: B(e),
    from: new c(i),
    mint: new c(a),
    custody: Me(e, a),
    authoritySigner: ne(e),
    custodySigner: fe(e),
    wormholeBridge: o,
    wormholeMessage: s,
    wormholeEmitter: d,
    wormholeSequence: u,
    wormholeFeeCollector: m,
    clock: h,
    rent: g,
    systemProgram: w,
    tokenProgram: M,
    wormholeProgram: new c(t)
  };
}
function ut(e, t, r, n, i, a, o, s, d, u, m, h) {
  const g = W(t, e).methods.transferNativeWithPayload(s, d, l.Buffer.from(u), m, l.Buffer.from(h), null);
  return g._ixFn(...g._args, {
    accounts: Bt(t, r, n, i, a, o),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function Bt(e, t, r, n, i, a, o) {
  const { wormholeBridge: s, wormholeMessage: d, wormholeEmitter: u, wormholeSequence: m, wormholeFeeCollector: h, clock: g, rent: w, systemProgram: p } = Se(e, t, r, n);
  return {
    payer: new c(r),
    config: B(e),
    from: new c(i),
    mint: new c(a),
    custody: Me(e, a),
    authoritySigner: ne(e),
    custodySigner: fe(e),
    wormholeBridge: s,
    wormholeMessage: d,
    wormholeEmitter: u,
    wormholeSequence: m,
    wormholeFeeCollector: h,
    clock: g,
    sender: new c(o === void 0 ? r : xe(o)),
    rent: w,
    systemProgram: p,
    tokenProgram: M,
    wormholeProgram: new c(t)
  };
}
function wr(e, t, r, n, i, a, o, s, d, u, m, h, g, w) {
  const p = W(t, e).methods.transferWrapped(u, m, h, l.Buffer.from(g), w);
  return p._ixFn(...p._args, {
    accounts: yr(t, r, n, i, a, o, s, d),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function yr(e, t, r, n, i, a, o, s) {
  const d = re(e, o, s), { wormholeBridge: u, wormholeMessage: m, wormholeEmitter: h, wormholeSequence: g, wormholeFeeCollector: w, clock: p, rent: C, systemProgram: R } = Se(e, t, r, n);
  return {
    payer: new c(r),
    config: B(e),
    from: new c(i),
    fromOwner: new c(a),
    mint: d,
    wrappedMeta: Y(e, d),
    authoritySigner: ne(e),
    wormholeBridge: u,
    wormholeMessage: m,
    wormholeEmitter: h,
    wormholeSequence: g,
    wormholeFeeCollector: w,
    clock: p,
    rent: C,
    systemProgram: R,
    wormholeProgram: new c(t),
    tokenProgram: M
  };
}
function Sr(e, t, r, n, i, a, o, s, d, u, m, h, g, w) {
  const p = W(t, e).methods.transferWrappedWithPayload(u, m, l.Buffer.from(h), g, l.Buffer.from(w), null);
  return p._ixFn(...p._args, {
    accounts: Rt(t, r, n, i, a, o, s, d),
    signers: void 0,
    remainingAccounts: void 0,
    preInstructions: void 0,
    postInstructions: void 0
  });
}
function Rt(e, t, r, n, i, a, o, s, d) {
  const u = re(e, o, s), { wormholeBridge: m, wormholeMessage: h, wormholeEmitter: g, wormholeSequence: w, wormholeFeeCollector: p, clock: C, rent: R, systemProgram: F } = Se(e, t, r, n);
  return {
    payer: new c(r),
    config: B(e),
    from: new c(i),
    fromOwner: new c(a),
    mint: u,
    wrappedMeta: Y(e, u),
    authoritySigner: ne(e),
    wormholeBridge: m,
    wormholeMessage: h,
    wormholeEmitter: g,
    wormholeSequence: w,
    wormholeFeeCollector: p,
    clock: C,
    sender: new c(d === void 0 ? r : xe(d)),
    rent: R,
    systemProgram: F,
    wormholeProgram: new c(t),
    tokenProgram: M
  };
}
function Ur(e, t, r) {
  const { wormholeEmitter: n, wormholeBridge: i, wormholeFeeCollector: a, wormholeSequence: o } = Zt(t, r);
  return {
    tokenBridgeConfig: B(t),
    tokenBridgeAuthoritySigner: ne(t),
    tokenBridgeCustodySigner: fe(t),
    tokenBridgeMintAuthority: ke(t),
    tokenBridgeSender: xe(e),
    tokenBridgeRedeemer: ze(e),
    wormholeBridge: i,
    tokenBridgeEmitter: n,
    wormholeFeeCollector: a,
    tokenBridgeSequence: o
  };
}
function Ar(e, t, r, n, i, a, o) {
  const s = Bt(t, r, n, i, a, o, e);
  return {
    payer: s.payer,
    tokenBridgeConfig: s.config,
    fromTokenAccount: s.from,
    mint: s.mint,
    tokenBridgeCustody: s.custody,
    tokenBridgeAuthoritySigner: s.authoritySigner,
    tokenBridgeCustodySigner: s.custodySigner,
    wormholeBridge: s.wormholeBridge,
    wormholeMessage: s.wormholeMessage,
    tokenBridgeEmitter: s.wormholeEmitter,
    tokenBridgeSequence: s.wormholeSequence,
    wormholeFeeCollector: s.wormholeFeeCollector,
    clock: s.clock,
    tokenBridgeSender: s.sender,
    rent: s.rent,
    systemProgram: s.systemProgram,
    tokenProgram: s.tokenProgram,
    wormholeProgram: s.wormholeProgram
  };
}
function Mr(e, t, r, n, i, a, o, s, d) {
  const u = Rt(t, r, n, i, a, d === void 0 ? e : d, o, s, e);
  return {
    payer: u.payer,
    tokenBridgeConfig: u.config,
    fromTokenAccount: u.from,
    fromTokenAccountOwner: u.fromOwner,
    tokenBridgeWrappedMint: u.mint,
    tokenBridgeWrappedMeta: u.wrappedMeta,
    tokenBridgeAuthoritySigner: u.authoritySigner,
    wormholeBridge: u.wormholeBridge,
    wormholeMessage: u.wormholeMessage,
    tokenBridgeEmitter: u.wormholeEmitter,
    tokenBridgeSequence: u.wormholeSequence,
    wormholeFeeCollector: u.wormholeFeeCollector,
    clock: u.clock,
    tokenBridgeSender: u.sender,
    rent: u.rent,
    systemProgram: u.systemProgram,
    tokenProgram: u.tokenProgram,
    wormholeProgram: u.wormholeProgram
  };
}
function Nr(e, t, r, n, i) {
  const a = new c(n.payload.token.address.toUint8Array()), o = new c(n.payload.to.address.toUint8Array());
  return {
    payer: new c(r),
    tokenBridgeConfig: B(e),
    vaa: H(t, l.Buffer.from(n.hash)),
    tokenBridgeClaim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    tokenBridgeForeignEndpoint: te(e, y(n.emitterChain), n.emitterAddress.toUint8Array()),
    toTokenAccount: new c(i),
    tokenBridgeRedeemer: ze(o),
    toFeesTokenAccount: new c(i),
    tokenBridgeCustody: Me(e, a),
    mint: a,
    tokenBridgeCustodySigner: fe(e),
    rent: O,
    systemProgram: E.programId,
    tokenProgram: M,
    wormholeProgram: new c(t)
  };
}
function Fr(e, t, r, n, i) {
  const a = re(e, y(n.payload.token.chain), n.payload.token.address.toUint8Array()), o = new c(n.payload.to.address.toUint8Array());
  return {
    payer: new c(r),
    tokenBridgeConfig: B(e),
    vaa: H(t, l.Buffer.from(n.hash)),
    tokenBridgeClaim: G(e, n.emitterAddress.toUint8Array(), y(n.emitterChain), n.sequence),
    tokenBridgeForeignEndpoint: te(e, y(n.emitterChain), n.emitterAddress.toUint8Array()),
    toTokenAccount: new c(i),
    tokenBridgeRedeemer: ze(o),
    toFeesTokenAccount: new c(i),
    tokenBridgeWrappedMint: a,
    tokenBridgeWrappedMeta: Y(e, a),
    tokenBridgeMintAuthority: ke(e),
    rent: O,
    systemProgram: E.programId,
    tokenProgram: M,
    wormholeProgram: new c(t)
  };
}
const kr = {
  version: "0.1.0",
  name: "token_bridge_relayer",
  constants: [
    {
      name: "SEED_PREFIX_BRIDGED",
      type: "bytes",
      value: "[98, 114, 105, 100, 103, 101, 100]"
    },
    {
      name: "SEED_PREFIX_TMP",
      type: "bytes",
      value: "[116, 109, 112]"
    },
    {
      name: "SWAP_RATE_PRECISION",
      type: "u32",
      value: "100_000_000"
    }
  ],
  instructions: [
    {
      name: "initialize",
      docs: [
        "This instruction is be used to generate your program's config.",
        "And for convenience, we will store Wormhole-related PDAs in the",
        "config so we can verify these accounts with a simple == constraint.",
        "# Arguments",
        "",
        "* `ctx`           - `Initialize` context",
        "* `fee_recipient` - Recipient of all relayer fees and swap proceeds",
        "* `assistant`     - Privileged key to manage certain accounts"
      ],
      accounts: [
        {
          name: "owner",
          isMut: !0,
          isSigner: !0,
          docs: ["Deployer of the program."]
        },
        {
          name: "senderConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Sender Config account, which saves program data useful for other",
            "instructions, specifically for outbound transfers. Also saves the payer",
            "of the [`initialize`](crate::initialize) instruction as the program's",
            "owner."
          ]
        },
        {
          name: "redeemerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Redeemer Config account, which saves program data useful for other",
            "instructions, specifically for inbound transfers. Also saves the payer",
            "of the [`initialize`](crate::initialize) instruction as the program's",
            "owner."
          ]
        },
        {
          name: "ownerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Owner config account, which saves the owner, assistant and",
            "pending owner keys. This account is used to manage the ownership of the",
            "program."
          ]
        },
        {
          name: "tokenBridgeEmitter",
          isMut: !1,
          isSigner: !1,
          docs: [
            "that holds data; it is purely just a signer for posting Wormhole",
            "messages on behalf of the Token Bridge program."
          ]
        },
        {
          name: "tokenBridgeSequence",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Token Bridge emitter's sequence account. Like with all Wormhole",
            "emitters, this account keeps track of the sequence number of the last",
            "posted message."
          ]
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1,
          docs: ["System program."]
        },
        {
          name: "programData",
          isMut: !0,
          isSigner: !1,
          docs: [
            "upgrade authority. We check this PDA address just in case there is another program that this",
            "deployer has deployed.",
            "",
            "NOTE: Set upgrade authority is scary because any public key can be used to set as the",
            "authority."
          ]
        },
        {
          name: "bpfLoaderUpgradeableProgram",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "feeRecipient",
          type: "publicKey"
        },
        {
          name: "assistant",
          type: "publicKey"
        }
      ]
    },
    {
      name: "registerForeignContract",
      docs: [
        "This instruction registers a new foreign contract (from another",
        "network) and saves the emitter information in a ForeignEmitter account.",
        "This instruction is owner-only, meaning that only the owner of the",
        "program (defined in the [Config] account) can add and update foreign",
        "contracts.",
        "",
        "# Arguments",
        "",
        "* `ctx`     - `RegisterForeignContract` context",
        "* `chain`   - Wormhole Chain ID",
        "* `address` - Wormhole Emitter Address",
        "* `relayer_fee` - Relayer fee scaled by the `relayer_fee_precision`"
      ],
      accounts: [
        {
          name: "owner",
          isMut: !0,
          isSigner: !0,
          docs: [
            "Owner of the program set in the [`SenderConfig`] account. Signer for",
            "creating [`ForeignContract`] account."
          ]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Sender Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Read-only."
          ]
        },
        {
          name: "foreignContract",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Foreign Contract account. Create this account if an emitter has not been",
            "registered yet for this Wormhole chain ID. If there already is a",
            "contract address saved in this account, overwrite it."
          ]
        },
        {
          name: "tokenBridgeForeignEndpoint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Token Bridge foreign endpoint. This account should really be one",
            "endpoint per chain, but Token Bridge's PDA allows for multiple",
            "endpoints for each chain. We store the proper endpoint for the",
            "emitter chain."
          ]
        },
        {
          name: "tokenBridgeProgram",
          isMut: !1,
          isSigner: !1,
          docs: ["Token Bridge program."]
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1,
          docs: ["System program."]
        }
      ],
      args: [
        {
          name: "chain",
          type: "u16"
        },
        {
          name: "address",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "relayerFee",
          type: "u64"
        }
      ]
    },
    {
      name: "registerToken",
      docs: [
        "This instruction registers a new token and saves the initial `swap_rate`",
        "and `max_native_token_amount` in a RegisteredToken account.",
        "This instruction is owner-only, meaning that only the owner of the",
        "program (defined in the [Config] account) can register a token.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `RegisterToken` context",
        "* `swap_rate`:",
        "- USD conversion rate scaled by the `swap_rate_precision`. For example,",
        "- if the conversion rate is $15 and the `swap_rate_precision` is",
        "- 1000000, the `swap_rate` should be set to 15000000.",
        "* `max_native_swap_amount`:",
        "- Maximum amount of native tokens that can be swapped for this token",
        "- on this chain."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !0,
          isSigner: !0,
          docs: [
            "Owner of the program set in the [`SenderConfig`] account. Signer for",
            "creating [`ForeignContract`] account."
          ]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Sender Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Read-only."
          ]
        },
        {
          name: "registeredToken",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Registered Token account. This account stores information about the",
            "token, including the swap rate and max native swap amount. Create this",
            "account if the mint has not been registered yet. Mutable."
          ]
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Mint info. This is the SPL token that will be bridged over to the",
            "foreign contract."
          ]
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1,
          docs: ["System program."]
        }
      ],
      args: [
        {
          name: "swapRate",
          type: "u64"
        },
        {
          name: "maxNativeSwapAmount",
          type: "u64"
        }
      ]
    },
    {
      name: "deregisterToken",
      docs: [
        "This instruction deregisters a token by closing the existing",
        "`RegisteredToken` account for a particular mint. This instruction is",
        "owner-only, meaning that only the owner of the program (defined in the",
        "[Config] account) can deregister a token."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: [
            "Owner of the program set in the [`SenderConfig`] account. Signer for",
            "closing [`RegisteredToken`] account."
          ]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Sender Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Read-only."
          ]
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Mint info. This is the SPL token that will be bridged over to the",
            "foreign contract."
          ]
        },
        {
          name: "registeredToken",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Registered Token account. This account stores information about the",
            "token, including the swap rate and max native swap amount. This account",
            "also determines if a mint is registered or not."
          ]
        }
      ],
      args: []
    },
    {
      name: "updateRelayerFee",
      docs: [
        "This instruction updates the `relayer_fee` in the `ForeignContract` account.",
        "The `relayer_fee` is scaled by the `relayer_fee_precision`. For example,",
        "if the `relayer_fee` is $15 and the `relayer_fee_precision` is 1000000,",
        "the `relayer_fee` should be set to 15000000. This instruction can",
        "only be called by the owner or assistant, which are defined in the",
        "[OwnerConfig] account.",
        "",
        "# Arguments",
        "",
        "* `ctx`   - `UpdateRelayerFee` context",
        "* `chain` - Wormhole Chain ID",
        "* `fee`   - Relayer fee scaled by the `relayer_fee_precision`"
      ],
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0,
          docs: ["Signer of the transaction. Must be the owner or assistant."]
        },
        {
          name: "ownerConfig",
          isMut: !1,
          isSigner: !1,
          docs: [
            "The owner_config is used when updating the swap rate",
            "so that the assistant key can be used in addition to the",
            "owner key."
          ]
        },
        {
          name: "foreignContract",
          isMut: !0,
          isSigner: !1,
          docs: [
            "This account holds the USD denominated relayer fee for the specified",
            "`chain`. This account is used to determine the cost of relaying",
            "a transfer to a target chain. If there already is a relayer fee",
            "saved in this account, overwrite it."
          ]
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1,
          docs: ["System program."]
        }
      ],
      args: [
        {
          name: "chain",
          type: "u16"
        },
        {
          name: "fee",
          type: "u64"
        }
      ]
    },
    {
      name: "updateRelayerFeePrecision",
      docs: [
        "This instruction updates the `relayer_fee_precision` in the",
        "`SenderConfig` and `RedeemerConfig` accounts. The `relayer_fee_precision`",
        "is used to scale the `relayer_fee`. This instruction is owner-only,",
        "meaning that only the owner of the program (defined in the [Config]",
        "account) can register a token.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `UpdatePrecision` context",
        "* `relayer_fee_precision` - Precision used to scale the relayer fee."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: [
            "Owner of the program set in the [`RedeemerConfig`] and [`SenderConfig`] account."
          ]
        },
        {
          name: "redeemerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Redeemer Config account. This program requires that the `owner`",
            "specified in the context equals the pubkey specified in this account.",
            "Mutable."
          ]
        },
        {
          name: "senderConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Sender Config account. This program requires that the `owner`",
            "specified in the context equals the pubkey specified in this account.",
            "Mutable. The `owner` check is redundant here, but we keep it as an",
            "extra protection for future changes to the context. Mutable."
          ]
        }
      ],
      args: [
        {
          name: "relayerFeePrecision",
          type: "u32"
        }
      ]
    },
    {
      name: "updateSwapRate",
      docs: [
        "This instruction updates the `swap_rate` in the `RegisteredToken`",
        "account. This instruction can only be called by the owner or",
        "assistant, which are defined in the [OwnerConfig] account.",
        "",
        "# Arguments",
        "",
        "* `ctx`       - `UpdateSwapRate` context",
        "* `swap_rate` - USD conversion rate for the specified token."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: [
            "The signer of the transaction. Must be the owner or assistant."
          ]
        },
        {
          name: "ownerConfig",
          isMut: !1,
          isSigner: !1,
          docs: [
            "The owner_config is used when updating the swap rate so that the",
            "assistant key can be used in additional to the owner key."
          ]
        },
        {
          name: "registeredToken",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Registered Token account. This account stores information about the",
            "token, including the swap rate and max native swap amount. The program",
            "will modify this account to update the swap rate. Mutable."
          ]
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Mint info. This is the SPL token that will be bridged over to the",
            "foreign contract."
          ]
        }
      ],
      args: [
        {
          name: "swapRate",
          type: "u64"
        }
      ]
    },
    {
      name: "updateMaxNativeSwapAmount",
      docs: [
        "This instruction updates the `max_native_swap_amount` in the",
        "`RegisteredToken` account. This instruction is owner-only,",
        "meaning that only the owner of the program (defined in the [Config]",
        "account) can register a token.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `UpdateMaxNativeSwapAmount` context",
        "* `max_native_swap_amount`:",
        "- Maximum amount of native tokens that can be swapped for this token",
        "- on this chain."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: [
            "Owner of the program set in the [`SenderConfig`] account. Signer for",
            "creating [`ForeignContract`] account."
          ]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Sender Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Read-only."
          ]
        },
        {
          name: "registeredToken",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Registered Token account. This account stores information about the",
            "token, including the swap rate and max native swap amount. The program",
            "will modify this account when the swap rate or max native swap amount",
            "changes. Mutable."
          ]
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Mint info. This is the SPL token that will be bridged over to the",
            "foreign contract."
          ]
        }
      ],
      args: [
        {
          name: "maxNativeSwapAmount",
          type: "u64"
        }
      ]
    },
    {
      name: "setPauseForTransfers",
      docs: [
        "This instruction updates the `paused` boolean in the `SenderConfig`",
        "account. This instruction is owner-only, meaning that only the owner",
        "of the program (defined in the [Config] account) can pause outbound",
        "transfers.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `PauseOutboundTransfers` context",
        "* `paused` - Boolean indicating whether outbound transfers are paused."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: ["Owner of the program set in the [`SenderConfig`] account."]
        },
        {
          name: "config",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Sender Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Mutable."
          ]
        }
      ],
      args: [
        {
          name: "paused",
          type: "bool"
        }
      ]
    },
    {
      name: "submitOwnershipTransferRequest",
      docs: [
        "This instruction sets the `pending_owner` field in the `OwnerConfig`",
        "account. This instruction is owner-only, meaning that only the owner",
        "of the program (defined in the [Config] account) can submit an",
        "ownership transfer request.",
        "",
        "# Arguments",
        "",
        "* `ctx`       - `ManageOwnership` context",
        "* `new_owner` - Pubkey of the pending owner."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: ["Owner of the program set in the [`OwnerConfig`] account."]
        },
        {
          name: "ownerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Owner Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Mutable."
          ]
        }
      ],
      args: [
        {
          name: "newOwner",
          type: "publicKey"
        }
      ]
    },
    {
      name: "confirmOwnershipTransferRequest",
      docs: [
        "This instruction confirms that the `pending_owner` is the signer of",
        "the transaction and updates the `owner` field in the `SenderConfig`,",
        "`RedeemerConfig`, and `OwnerConfig` accounts."
      ],
      accounts: [
        {
          name: "pendingOwner",
          isMut: !1,
          isSigner: !0,
          docs: [
            "Must be the pending owner of the program set in the [`OwnerConfig`]",
            "account."
          ]
        },
        {
          name: "ownerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Owner Config account. This program requires that the `pending_owner`",
            "specified in the context equals the pubkey specified in this account."
          ]
        },
        {
          name: "senderConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Sender Config account. This instruction will update the `owner`",
            "specified in this account to the `pending_owner` specified in the",
            "[`OwnerConfig`] account. Mutable."
          ]
        },
        {
          name: "redeemerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Redeemer Config account. This instruction will update the `owner`",
            "specified in this account to the `pending_owner` specified in the",
            "[`OwnerConfig`] account. Mutable."
          ]
        }
      ],
      args: []
    },
    {
      name: "cancelOwnershipTransferRequest",
      docs: [
        "This instruction cancels the ownership transfer request by setting",
        "the `pending_owner` field in the `OwnerConfig` account to `None`.",
        "This instruction is owner-only, meaning that only the owner of the",
        "program (defined in the [Config] account) can cancel an ownership",
        "transfer request."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: ["Owner of the program set in the [`OwnerConfig`] account."]
        },
        {
          name: "ownerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Owner Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Mutable."
          ]
        }
      ],
      args: []
    },
    {
      name: "updateAssistant",
      docs: [
        "This instruction updates the `assistant` field in the `OwnerConfig`",
        "account. This instruction is owner-only, meaning that only the owner",
        "of the program (defined in the [Config] account) can update the",
        "assistant.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `ManageOwnership` context",
        "* `new_assistant` - Pubkey of the new assistant."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: ["Owner of the program set in the [`OwnerConfig`] account."]
        },
        {
          name: "ownerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Owner Config account. This program requires that the `owner` specified",
            "in the context equals the pubkey specified in this account. Mutable."
          ]
        }
      ],
      args: [
        {
          name: "newAssistant",
          type: "publicKey"
        }
      ]
    },
    {
      name: "updateFeeRecipient",
      docs: [
        "This instruction updates the `fee_recipient` field in the `RedeemerConfig`",
        "account. This instruction is owner-only, meaning that only the owner",
        "of the program (defined in the [Config] account) can update the",
        "fee recipient.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `UpdateFeeRecipient` context",
        "* `new_fee_recipient` - Pubkey of the new fee recipient."
      ],
      accounts: [
        {
          name: "owner",
          isMut: !1,
          isSigner: !0,
          docs: ["Owner of the program set in the [`RedeemerConfig`] account."]
        },
        {
          name: "redeemerConfig",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Redeemer Config account, which saves program data useful for other",
            "instructions, specifically for inbound transfers. Also saves the payer",
            "of the [`initialize`](crate::initialize) instruction as the program's",
            "owner."
          ]
        }
      ],
      args: [
        {
          name: "newFeeRecipient",
          type: "publicKey"
        }
      ]
    },
    {
      name: "transferNativeTokensWithRelay",
      docs: [
        "This instruction is used to transfer native tokens from Solana to a",
        "foreign blockchain. The user can optionally specify a",
        "`to_native_token_amount` to swap some of the tokens for the native",
        "asset on the target chain. For a fee, an off-chain relayer will redeem",
        "the transfer on the target chain. If the user is transferring native",
        "SOL, the contract will automatically wrap the lamports into a WSOL.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `TransferNativeWithRelay` context",
        "* `amount` - Amount of tokens to send",
        "* `to_native_token_amount`:",
        "- Amount of tokens to swap for native assets on the target chain",
        "* `recipient_chain` - Chain ID of the target chain",
        "* `recipient_address` - Address of the target wallet on the target chain",
        "* `batch_id` - Nonce of Wormhole message",
        "* `wrap_native` - Whether to wrap native SOL"
      ],
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0,
          docs: [
            "Payer will pay Wormhole fee to transfer tokens and create temporary",
            "token account."
          ]
        },
        {
          name: "payerSequence",
          isMut: !0,
          isSigner: !1,
          docs: ["Used to keep track of payer's Wormhole sequence number."]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Sender Config account. Acts as the signer for the Token Bridge token",
            "transfer. Read-only."
          ]
        },
        {
          name: "foreignContract",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Foreign Contract account. Send tokens to the contract specified in this",
            "account. Funnily enough, the Token Bridge program does not have any",
            "requirements for outbound transfers for the recipient chain to be",
            "registered. This account provides extra protection against sending",
            "tokens to an unregistered Wormhole chain ID. Read-only."
          ]
        },
        {
          name: "mint",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Mint info. This is the SPL token that will be bridged over to the",
            "foreign contract. Mutable."
          ]
        },
        {
          name: "fromTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Payer's associated token account. We may want to make this a generic",
            "token account in the future."
          ]
        },
        {
          name: "registeredToken",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tmpTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Program's temporary token account. This account is created before the",
            "instruction is invoked to temporarily take custody of the payer's",
            "tokens. When the tokens are finally bridged out, the token account",
            "will have zero balance and can be closed."
          ]
        },
        {
          name: "tokenBridgeConfig",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeCustody",
          isMut: !0,
          isSigner: !1,
          docs: [
            "account that holds this mint's balance. This account needs to be",
            "unchecked because a token account may not have been created for this",
            "mint yet. Mutable."
          ]
        },
        {
          name: "tokenBridgeAuthoritySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeCustodySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !1,
          docs: [
            "tokens transferred in this account for our program. Mutable."
          ]
        },
        {
          name: "tokenBridgeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "amount",
          type: "u64"
        },
        {
          name: "toNativeTokenAmount",
          type: "u64"
        },
        {
          name: "recipientChain",
          type: "u16"
        },
        {
          name: "recipientAddress",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "batchId",
          type: "u32"
        },
        {
          name: "wrapNative",
          type: "bool"
        }
      ]
    },
    {
      name: "transferWrappedTokensWithRelay",
      docs: [
        "This instruction is used to transfer wrapped tokens from Solana to a",
        "foreign blockchain. The user can optionally specify a",
        "`to_native_token_amount` to swap some of the tokens for the native",
        "assets on the target chain. For a fee, an off-chain relayer will redeem",
        "the transfer on the target chain. This instruction should only be called",
        "when the user is transferring a wrapped token.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `TransferWrappedWithRelay` context",
        "* `amount` - Amount of tokens to send",
        "* `to_native_token_amount`:",
        "- Amount of tokens to swap for native assets on the target chain",
        "* `recipient_chain` - Chain ID of the target chain",
        "* `recipient_address` - Address of the target wallet on the target chain",
        "* `batch_id` - Nonce of Wormhole message"
      ],
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0,
          docs: ["Payer will pay Wormhole fee to transfer tokens."]
        },
        {
          name: "payerSequence",
          isMut: !0,
          isSigner: !1,
          docs: ["Used to keep track of payer's Wormhole sequence number."]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Sender Config account. Acts as the Token Bridge sender PDA. Mutable."
          ]
        },
        {
          name: "foreignContract",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Foreign Contract account. Send tokens to the contract specified in this",
            "account. Funnily enough, the Token Bridge program does not have any",
            "requirements for outbound transfers for the recipient chain to be",
            "registered. This account provides extra protection against sending",
            "tokens to an unregistered Wormhole chain ID. Read-only."
          ]
        },
        {
          name: "tokenBridgeWrappedMint",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Token Bridge wrapped mint info. This is the SPL token that will be",
            "bridged to the foreign contract. The wrapped mint PDA must agree",
            "with the native token's metadata. Mutable."
          ]
        },
        {
          name: "fromTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Payer's associated token account. We may want to make this a generic",
            "token account in the future."
          ]
        },
        {
          name: "registeredToken",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tmpTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Program's temporary token account. This account is created before the",
            "instruction is invoked to temporarily take custody of the payer's",
            "tokens. When the tokens are finally bridged out, the token account",
            "will have zero balance and can be closed."
          ]
        },
        {
          name: "tokenBridgeWrappedMeta",
          isMut: !1,
          isSigner: !1,
          docs: [
            "about the token from its native chain:",
            "* Wormhole Chain ID",
            "* Token's native contract address",
            "* Token's native decimals"
          ]
        },
        {
          name: "tokenBridgeConfig",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeAuthoritySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeBridge",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeMessage",
          isMut: !0,
          isSigner: !1,
          docs: ["tokens transferred in this account."]
        },
        {
          name: "tokenBridgeEmitter",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeSequence",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeFeeCollector",
          isMut: !0,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "clock",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "amount",
          type: "u64"
        },
        {
          name: "toNativeTokenAmount",
          type: "u64"
        },
        {
          name: "recipientChain",
          type: "u16"
        },
        {
          name: "recipientAddress",
          type: {
            array: ["u8", 32]
          }
        },
        {
          name: "batchId",
          type: "u32"
        }
      ]
    },
    {
      name: "completeNativeTransferWithRelay",
      docs: [
        "This instruction is used to redeem token transfers from foreign emitters.",
        "It takes custody of the released native tokens and sends the tokens to the",
        "encoded `recipient`. It pays the `fee_recipient` in the token",
        "denomination. If requested by the user, it will perform a swap with the",
        "off-chain relayer to provide the user with lamports. If the token",
        "being transferred is WSOL, the contract will unwrap the WSOL and send",
        "the lamports to the recipient and pay the relayer in lamports.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `CompleteNativeWithRelay` context",
        "* `vaa_hash` - Hash of the VAA that triggered the transfer"
      ],
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0,
          docs: [
            "Payer will pay Wormhole fee to transfer tokens and create temporary",
            "token account."
          ]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Redeemer Config account. Acts as the Token Bridge redeemer, which signs",
            "for the complete transfer instruction. Read-only."
          ]
        },
        {
          name: "feeRecipientTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Fee recipient's token account. Must be an associated token account. Mutable."
          ]
        },
        {
          name: "foreignContract",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Foreign Contract account. The registered contract specified in this",
            "account must agree with the target address for the Token Bridge's token",
            "transfer. Read-only."
          ]
        },
        {
          name: "mint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Mint info. This is the SPL token that will be bridged over from the",
            "foreign contract. This must match the token address specified in the",
            "signed Wormhole message. Read-only."
          ]
        },
        {
          name: "recipientTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Recipient associated token account. The recipient authority check",
            "is necessary to ensure that the recipient is the intended recipient",
            "of the bridged tokens. Mutable."
          ]
        },
        {
          name: "recipient",
          isMut: !0,
          isSigner: !1,
          docs: [
            "transaction. This instruction verifies that the recipient key",
            "passed in this context matches the intended recipient in the vaa."
          ]
        },
        {
          name: "registeredToken",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "nativeRegisteredToken",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tmpTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Program's temporary token account. This account is created before the",
            "instruction is invoked to temporarily take custody of the payer's",
            "tokens. When the tokens are finally bridged in, the tokens will be",
            "transferred to the destination token accounts. This account will have",
            "zero balance and can be closed."
          ]
        },
        {
          name: "tokenBridgeConfig",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Verified Wormhole message account. The Wormhole program verified",
            "signatures and posted the account data here. Read-only."
          ]
        },
        {
          name: "tokenBridgeClaim",
          isMut: !0,
          isSigner: !1,
          docs: [
            "is true if the bridged assets have been claimed. If the transfer has",
            "not been redeemed, this account will not exist yet.",
            "",
            "NOTE: The Token Bridge program's claim account is only initialized when",
            "a transfer is redeemed (and the boolean value `true` is written as",
            "its data).",
            "",
            "The Token Bridge program will automatically fail if this transfer",
            "is redeemed again. But we choose to short-circuit the failure as the",
            "first evaluation of this instruction."
          ]
        },
        {
          name: "tokenBridgeForeignEndpoint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "endpoint per chain, but the PDA allows for multiple endpoints for each",
            "chain! We store the proper endpoint for the emitter chain."
          ]
        },
        {
          name: "tokenBridgeCustody",
          isMut: !0,
          isSigner: !1,
          docs: ["account that holds this mint's balance."]
        },
        {
          name: "tokenBridgeCustodySigner",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "vaaHash",
          type: {
            array: ["u8", 32]
          }
        }
      ]
    },
    {
      name: "completeWrappedTransferWithRelay",
      docs: [
        "This instruction is used to redeem token transfers from foreign emitters.",
        "It takes custody of the minted wrapped tokens and sends the tokens to the",
        "encoded `recipient`. It pays the `fee_recipient` in the wrapped-token",
        "denomination. If requested by the user, it will perform a swap with the",
        "off-chain relayer to provide the user with lamports.",
        "",
        "# Arguments",
        "",
        "* `ctx` - `CompleteWrappedWithRelay` context",
        "* `vaa_hash` - Hash of the VAA that triggered the transfer"
      ],
      accounts: [
        {
          name: "payer",
          isMut: !0,
          isSigner: !0,
          docs: [
            "Payer will pay Wormhole fee to transfer tokens and create temporary",
            "token account."
          ]
        },
        {
          name: "config",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Redeemer Config account. Acts as the Token Bridge redeemer, which signs",
            "for the complete transfer instruction. Read-only."
          ]
        },
        {
          name: "feeRecipientTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Fee recipient's token account. Must be an associated token account. Mutable."
          ]
        },
        {
          name: "foreignContract",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Foreign Contract account. The registered contract specified in this",
            "account must agree with the target address for the Token Bridge's token",
            "transfer. Read-only."
          ]
        },
        {
          name: "tokenBridgeWrappedMint",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Token Bridge wrapped mint info. This is the SPL token that will be",
            "bridged from the foreign contract. The wrapped mint PDA must agree",
            "with the native token's metadata in the wormhole message. Mutable."
          ]
        },
        {
          name: "recipientTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Recipient associated token account. The recipient authority check",
            "is necessary to ensure that the recipient is the intended recipient",
            "of the bridged tokens. Mutable."
          ]
        },
        {
          name: "recipient",
          isMut: !0,
          isSigner: !1,
          docs: [
            "transaction. This instruction verifies that the recipient key",
            "passed in this context matches the intended recipient in the vaa."
          ]
        },
        {
          name: "registeredToken",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "nativeRegisteredToken",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tmpTokenAccount",
          isMut: !0,
          isSigner: !1,
          docs: [
            "Program's temporary token account. This account is created before the",
            "instruction is invoked to temporarily take custody of the payer's",
            "tokens. When the tokens are finally bridged in, the tokens will be",
            "transferred to the destination token accounts. This account will have",
            "zero balance and can be closed."
          ]
        },
        {
          name: "tokenBridgeWrappedMeta",
          isMut: !1,
          isSigner: !1,
          docs: [
            "about the token from its native chain:",
            "* Wormhole Chain ID",
            "* Token's native contract address",
            "* Token's native decimals"
          ]
        },
        {
          name: "tokenBridgeConfig",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "vaa",
          isMut: !1,
          isSigner: !1,
          docs: [
            "Verified Wormhole message account. The Wormhole program verified",
            "signatures and posted the account data here. Read-only."
          ]
        },
        {
          name: "tokenBridgeClaim",
          isMut: !0,
          isSigner: !1,
          docs: [
            "is true if the bridged assets have been claimed. If the transfer has",
            "not been redeemed, this account will not exist yet.",
            "",
            "NOTE: The Token Bridge program's claim account is only initialized when",
            "a transfer is redeemed (and the boolean value `true` is written as",
            "its data).",
            "",
            "The Token Bridge program will automatically fail if this transfer",
            "is redeemed again. But we choose to short-circuit the failure as the",
            "first evaluation of this instruction."
          ]
        },
        {
          name: "tokenBridgeForeignEndpoint",
          isMut: !1,
          isSigner: !1,
          docs: [
            "endpoint per chain, but the PDA allows for multiple endpoints for each",
            "chain! We store the proper endpoint for the emitter chain."
          ]
        },
        {
          name: "tokenBridgeMintAuthority",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "wormholeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenBridgeProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "systemProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "tokenProgram",
          isMut: !1,
          isSigner: !1
        },
        {
          name: "rent",
          isMut: !1,
          isSigner: !1
        }
      ],
      args: [
        {
          name: "vaaHash",
          type: {
            array: ["u8", 32]
          }
        }
      ]
    }
  ],
  accounts: [
    {
      name: "foreignContract",
      docs: ["Foreign emitter account data."],
      type: {
        kind: "struct",
        fields: [
          {
            name: "chain",
            docs: ["Emitter chain. Cannot equal `1` (Solana's Chain ID)."],
            type: "u16"
          },
          {
            name: "address",
            docs: ["Emitter address. Cannot be zero address."],
            type: {
              array: ["u8", 32]
            }
          },
          {
            name: "tokenBridgeForeignEndpoint",
            docs: ["Token Bridge program's foreign endpoint account key."],
            type: "publicKey"
          },
          {
            name: "fee",
            docs: [
              "The fee that is paid to the `fee_recipient` upon redeeming a transfer.",
              "This value is set in terms of USD and scaled by the `relayer_fee_precision`.",
              "For example, if the `relayer_fee_precision` is `100000000` and the intended",
              "fee is $5, then the `fee` value should be `500000000`."
            ],
            type: "u64"
          }
        ]
      }
    },
    {
      name: "ownerConfig",
      docs: ["Owner account data."],
      type: {
        kind: "struct",
        fields: [
          {
            name: "owner",
            docs: ["Program's owner."],
            type: "publicKey"
          },
          {
            name: "assistant",
            docs: [
              "Program's assistant. Can be used to update the relayer fee and swap rate."
            ],
            type: "publicKey"
          },
          {
            name: "pendingOwner",
            docs: [
              "Intermediate storage for the pending owner. Is used to transfer ownership."
            ],
            type: {
              option: "publicKey"
            }
          }
        ]
      }
    },
    {
      name: "redeemerConfig",
      type: {
        kind: "struct",
        fields: [
          {
            name: "owner",
            docs: ["Program's owner."],
            type: "publicKey"
          },
          {
            name: "bump",
            docs: ["PDA bump."],
            type: "u8"
          },
          {
            name: "relayerFeePrecision",
            docs: ["Relayer fee and swap rate precision."],
            type: "u32"
          },
          {
            name: "feeRecipient",
            docs: ["Recipient of all relayer fees and swap proceeds."],
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "registeredToken",
      docs: ["Registered token account data."],
      type: {
        kind: "struct",
        fields: [
          {
            name: "swapRate",
            docs: [
              "Token swap rate. The swap rate is the USD conversion rate of the token."
            ],
            type: "u64"
          },
          {
            name: "maxNativeSwapAmount",
            docs: [
              "Maximum amount of native SOL the contract will swap for each transfer."
            ],
            type: "u64"
          }
        ]
      }
    },
    {
      name: "senderConfig",
      type: {
        kind: "struct",
        fields: [
          {
            name: "owner",
            docs: ["Program's owner."],
            type: "publicKey"
          },
          {
            name: "bump",
            docs: ["PDA bump."],
            type: "u8"
          },
          {
            name: "tokenBridge",
            docs: ["Token Bridge program's relevant addresses."],
            type: {
              defined: "OutboundTokenBridgeAddresses"
            }
          },
          {
            name: "relayerFeePrecision",
            docs: ["Relayer fee and swap rate precision."],
            type: "u32"
          },
          {
            name: "paused",
            docs: ["Boolean indicating whether outbound transfers are paused."],
            type: "bool"
          }
        ]
      }
    },
    {
      name: "signerSequence",
      type: {
        kind: "struct",
        fields: [
          {
            name: "value",
            type: "u64"
          }
        ]
      }
    }
  ],
  types: [
    {
      name: "OutboundTokenBridgeAddresses",
      type: {
        kind: "struct",
        fields: [
          {
            name: "sequence",
            type: "publicKey"
          }
        ]
      }
    },
    {
      name: "TokenBridgeRelayerMessage",
      docs: [
        "Expected message types for this program. Only valid payloads are:",
        "* `TransferWithRelay`: Payload ID == 1.",
        "",
        "Payload IDs are encoded as u8."
      ],
      type: {
        kind: "enum",
        variants: [
          {
            name: "TransferWithRelay",
            fields: [
              {
                name: "target_relayer_fee",
                type: "u64"
              },
              {
                name: "to_native_token_amount",
                type: "u64"
              },
              {
                name: "recipient",
                type: {
                  array: ["u8", 32]
                }
              }
            ]
          }
        ]
      }
    }
  ],
  errors: [
    {
      code: 6e3,
      name: "InvalidWormholeBridge",
      msg: "InvalidWormholeBridge"
    },
    {
      code: 6001,
      name: "InvalidWormholeFeeCollector",
      msg: "InvalidWormholeFeeCollector"
    },
    {
      code: 6002,
      name: "OwnerOnly",
      msg: "OwnerOnly"
    },
    {
      code: 6003,
      name: "OutboundTransfersPaused",
      msg: "OutboundTransfersPaused"
    },
    {
      code: 6004,
      name: "OwnerOrAssistantOnly",
      msg: "OwnerOrAssistantOnly"
    },
    {
      code: 6005,
      name: "NotPendingOwner",
      msg: "NotPendingOwner"
    },
    {
      code: 6006,
      name: "AlreadyTheOwner",
      msg: "AlreadyTheOwner"
    },
    {
      code: 6007,
      name: "AlreadyTheAssistant",
      msg: "AlreadyTheAssistant"
    },
    {
      code: 6008,
      name: "AlreadyTheFeeRecipient",
      msg: "AlreadyTheFeeRecipient"
    },
    {
      code: 6009,
      name: "BumpNotFound",
      msg: "BumpNotFound"
    },
    {
      code: 6010,
      name: "FailedToMakeImmutable",
      msg: "FailedToMakeImmutable"
    },
    {
      code: 6011,
      name: "InvalidForeignContract",
      msg: "InvalidForeignContract"
    },
    {
      code: 6012,
      name: "ZeroBridgeAmount",
      msg: "ZeroBridgeAmount"
    },
    {
      code: 6013,
      name: "InvalidToNativeAmount",
      msg: "InvalidToNativeAmount"
    },
    {
      code: 6014,
      name: "NativeMintRequired",
      msg: "NativeMintRequired"
    },
    {
      code: 6015,
      name: "SwapsNotAllowedForNativeMint",
      msg: "SwapsNotAllowedForNativeMint"
    },
    {
      code: 6016,
      name: "InvalidTokenBridgeConfig",
      msg: "InvalidTokenBridgeConfig"
    },
    {
      code: 6017,
      name: "InvalidTokenBridgeAuthoritySigner",
      msg: "InvalidTokenBridgeAuthoritySigner"
    },
    {
      code: 6018,
      name: "InvalidTokenBridgeCustodySigner",
      msg: "InvalidTokenBridgeCustodySigner"
    },
    {
      code: 6019,
      name: "InvalidTokenBridgeEmitter",
      msg: "InvalidTokenBridgeEmitter"
    },
    {
      code: 6020,
      name: "InvalidTokenBridgeSequence",
      msg: "InvalidTokenBridgeSequence"
    },
    {
      code: 6021,
      name: "InvalidRecipient",
      msg: "InvalidRecipient"
    },
    {
      code: 6022,
      name: "InvalidTransferToChain",
      msg: "InvalidTransferToChain"
    },
    {
      code: 6023,
      name: "InvalidTransferTokenChain",
      msg: "InvalidTransferTokenChain"
    },
    {
      code: 6024,
      name: "InvalidPrecision",
      msg: "InvalidPrecision"
    },
    {
      code: 6025,
      name: "InvalidTransferToAddress",
      msg: "InvalidTransferToAddress"
    },
    {
      code: 6026,
      name: "AlreadyRedeemed",
      msg: "AlreadyRedeemed"
    },
    {
      code: 6027,
      name: "InvalidTokenBridgeForeignEndpoint",
      msg: "InvalidTokenBridgeForeignEndpoint"
    },
    {
      code: 6028,
      name: "InvalidTokenBridgeMintAuthority",
      msg: "InvalidTokenBridgeMintAuthority"
    },
    {
      code: 6029,
      name: "InvalidPublicKey",
      msg: "InvalidPublicKey"
    },
    {
      code: 6030,
      name: "ZeroSwapRate",
      msg: "ZeroSwapRate"
    },
    {
      code: 6031,
      name: "TokenNotRegistered",
      msg: "TokenNotRegistered"
    },
    {
      code: 6032,
      name: "ChainNotRegistered",
      msg: "ChainNotRegistered"
    },
    {
      code: 6033,
      name: "TokenAlreadyRegistered",
      msg: "TokenAlreadyRegistered"
    },
    {
      code: 6034,
      name: "FeeCalculationError",
      msg: "TokenFeeCalculationError"
    },
    {
      code: 6035,
      name: "InvalidSwapCalculation",
      msg: "InvalidSwapCalculation"
    },
    {
      code: 6036,
      name: "InsufficientFunds",
      msg: "InsufficientFunds"
    }
  ]
};
function De(e, t) {
  return new lt(kr, new c(e), { connection: t });
}
function Et(e, t) {
  return T([l.Buffer.from("seq"), new c(t).toBuffer()], e);
}
async function br(e, t, r, n, i, a, o, s, d, u, m, h) {
  const { methods: { transferNativeTokensWithRelay: g }, account: { signerSequence: w } } = De(t, e), p = Et(t, r), C = await w.fetch(p).then(({ value: x }) => x).catch((x) => {
    var q;
    if ((q = x.message) != null && q.includes("Account does not exist"))
      return new v(0);
    throw x;
  }), R = bt(t, r, C), F = We(new c(a), new c(r)), I = kt(t, a), ie = Ar(t, n, i, r, R, F, a);
  return g(new v(o.toString()), new v(s.toString()), y(u), [...d], m, h).accounts({
    config: Mt(t),
    payerSequence: p,
    foreignContract: Ie(t, u),
    registeredToken: Ue(t, a),
    tmpTokenAccount: I,
    tokenBridgeProgram: new c(n),
    ...ie
  }).instruction();
}
async function Tr(e, t, r, n, i, a, o, s, d, u, m) {
  const { methods: { transferWrappedTokensWithRelay: h }, account: { signerSequence: g } } = De(t, e), w = Et(t, r), p = await g.fetch(w).then(({ value: q }) => q).catch((q) => {
    var Ve;
    if ((Ve = q.message) != null && Ve.includes("Account does not exist"))
      return new v(0);
    throw q;
  }), C = bt(t, r, p), R = We(new c(a), new c(r)), { chain: F, tokenAddress: I } = await me(e, n, a), ie = kt(t, a), x = Mr(t, n, i, r, C, R, F, I);
  return h(new v(o.toString()), new v(s.toString()), y(u), [...d], m).accounts({
    config: Mt(t),
    payerSequence: w,
    foreignContract: Ie(t, u),
    registeredToken: Ue(t, new c(a)),
    tmpTokenAccount: ie,
    tokenBridgeProgram: new c(n),
    ...x
  }).instruction();
}
class qe {
  constructor(t, r, n, i) {
    f(this, "network");
    f(this, "chain");
    f(this, "connection");
    f(this, "contracts");
    f(this, "chainId");
    f(this, "coreBridge");
    f(this, "tokenBridge");
    this.network = t, this.chain = r, this.connection = n, this.contracts = i, this.chainId = y(r);
    const a = i.tokenBridge;
    if (!a)
      throw new Error(`TokenBridge contract Address for chain ${r} not found`);
    this.tokenBridge = W(a, n), this.coreBridge = new Gt(t, r, n, i);
  }
  static async fromRpc(t, r) {
    const [n, i] = await oe.chainFromRpc(t), a = r[i];
    if (a.network !== n)
      throw new Error(`Network mismatch for chain ${i}: ${a.network} != ${n}`);
    return new qe(n, i, t, a.contracts);
  }
  async isWrappedAsset(t) {
    return me(this.connection, this.tokenBridge.programId, new k(t).toUint8Array()).catch((r) => null).then((r) => r != null);
  }
  async getOriginalAsset(t) {
    if (!await this.isWrappedAsset(t))
      throw Ce(t.toString());
    const r = new k(t).toUint8Array(), n = new c(r);
    try {
      const i = await me(this.connection, this.tokenBridge.programId, r);
      return i === null ? {
        chain: this.chain,
        address: new k(n.toBytes()).toUniversalAddress()
      } : {
        chain: Ht(i.chain),
        address: new dt(new Uint8Array(i.tokenAddress))
      };
    } catch {
      throw Ce(t.toString());
    }
  }
  async getTokenUniversalAddress(t) {
    return new k(t).toUniversalAddress();
  }
  async getTokenNativeAddress(t, r) {
    return new k(r).toNative();
  }
  async hasWrappedAsset(t) {
    try {
      return await this.getWrappedAsset(t), !0;
    } catch {
    }
    return !1;
  }
  async getWrappedAsset(t) {
    if ($(t.address))
      throw new Error("Native cannot be a wrapped asset");
    const r = re(this.tokenBridge.programId, y(t.chain), t.address.toUniversalAddress().toUint8Array());
    try {
      return await me(this.connection, this.tokenBridge.programId, r), Be(this.chain, r.toBase58());
    } catch {
    }
    throw Ce(`${r}: ${t.address.toUniversalAddress().toString()}`);
  }
  async isTransferCompleted(t) {
    return Yt(this.connection, this.tokenBridge.programId, t.emitterAddress.toUint8Array(), y(t.emitterChain), t.sequence, this.connection.commitment).catch(() => !1);
  }
  async getWrappedNative() {
    return Be(this.chain, D.toBase58());
  }
  async *createAttestation(t, r) {
    if (!r)
      throw new Error("Payer required to create attestation");
    const n = new k(r).unwrap(), i = 0, a = await this.coreBridge.getMessageFee(), o = Xt(this.coreBridge.coreBridge.programId, n, a), s = ae.generate(), d = or(this.connection, this.tokenBridge.programId, this.coreBridge.address, n, new k(t).unwrap(), s.publicKey, i), u = new L().add(o, d);
    u.feePayer = n, yield this.createUnsignedTx({ transaction: u, signers: [s] }, "Solana.AttestToken");
  }
  async *submitAttestation(t, r) {
    if (!r)
      throw new Error("Payer required to create attestation");
    const n = new k(r).unwrap();
    yield* this.coreBridge.postVaa(n, t);
    const i = new L().add(lr(this.connection, this.tokenBridge.programId, this.coreBridge.address, n, t));
    i.feePayer = n, yield this.createUnsignedTx({ transaction: i }, "Solana.CreateWrapped");
  }
  async transferSol(t, r, n, i) {
    const a = new k(t).unwrap(), o = a, s = r.address.toUniversalAddress().toUint8Array(), d = y(r.chain), u = 0, m = 0n, h = ae.generate(), g = ae.generate(), w = await Qe(this.connection), p = E.createAccount({
      fromPubkey: o,
      newAccountPubkey: g.publicKey,
      lamports: w,
      //spl token accounts need rent exemption
      space: P,
      programId: M
    }), C = E.transfer({
      fromPubkey: o,
      lamports: n,
      toPubkey: g.publicKey
    }), R = nt(g.publicKey, D, o), F = it(this.tokenBridge.programId, g.publicKey, o, n), I = i ? ut(this.connection, this.tokenBridge.programId, this.coreBridge.address, a, h.publicKey, g.publicKey, D, u, n, s, d, i) : ct(this.connection, this.tokenBridge.programId, this.coreBridge.address, a, h.publicKey, g.publicKey, D, u, n, m, s, d), ie = Xe(
      g.publicKey,
      //account to close
      o,
      //Remaining funds destination
      o
    ), x = new L();
    return x.feePayer = o, x.add(p, C, R, F, I, ie), this.createUnsignedTx({ transaction: x, signers: [h, g] }, "TokenBridge.TransferNative");
  }
  async *transfer(t, r, n, i, a) {
    if ($(n)) {
      yield await this.transferSol(t, r, i, a);
      return;
    }
    const o = new k(n).unwrap(), s = new k(t).unwrap(), d = await Je(o, s), u = r.address.toUniversalAddress().toUint8Array(), m = y(r.chain), h = 0, g = 0n, w = !await this.isWrappedAsset(n), p = ae.generate();
    let C;
    if (w)
      C = a ? ut(this.connection, this.tokenBridge.programId, this.coreBridge.address, s, p.publicKey, d, o, h, i, u, m, a) : ct(this.connection, this.tokenBridge.programId, this.coreBridge.address, s, p.publicKey, d, o, h, i, g, u, m);
    else {
      const I = await this.getOriginalAsset(n);
      if ($(I.address))
        throw new Error("Native cannot be an original asset");
      C = a ? Sr(this.connection, this.tokenBridge.programId, this.coreBridge.address, s, p.publicKey, d, s, y(I.chain), I.address.toUint8Array(), h, i, u, m, a) : wr(this.connection, this.tokenBridge.programId, this.coreBridge.address, s, p.publicKey, d, s, y(I.chain), I.address.toUint8Array(), h, i, g, u, m);
    }
    const R = it(this.tokenBridge.programId, d, s, i), F = new L().add(R, C);
    F.feePayer = s, yield this.createUnsignedTx({ transaction: F, signers: [p] }, "TokenBridge.TransferTokens");
  }
  async *redeemAndUnwrap(t, r) {
    const n = new k(t).unwrap(), i = new k(r.payload.to.address).unwrap(), a = await an(this.connection, D).then((p) => r.payload.token.amount * BigInt(Math.pow(10, p.decimals - 8))), o = await Qe(this.connection), s = ae.generate(), d = ot(this.connection, this.tokenBridge.programId, this.coreBridge.address, n, r), u = E.createAccount({
      fromPubkey: n,
      newAccountPubkey: s.publicKey,
      lamports: o,
      //spl token accounts need rent exemption
      space: P,
      programId: M
    }), m = nt(s.publicKey, D, n), h = Ln(i, s.publicKey, n, a.valueOf()), g = Xe(
      s.publicKey,
      //account to close
      n,
      //Remaining funds destination
      n
    ), w = new L();
    w.feePayer = n, w.add(d, u, m, h, g), yield this.createUnsignedTx({ transaction: w, signers: [s] }, "TokenBridge.RedeemAndUnwrap");
  }
  async *createAta(t, r) {
    const n = new k(t).unwrap(), i = new k(r).unwrap(), a = await Je(i, n);
    if (await this.connection.getAccountInfo(a) === null) {
      const s = new L().add(At(n, a, n, i));
      s.feePayer = n, yield this.createUnsignedTx({ transaction: s }, "Redeem.CreateATA");
    }
  }
  async *redeem(t, r, n = !0) {
    const i = r.payload.token.chain === this.chain ? r.payload.token.address : (await this.getWrappedAsset(r.payload.token)).toUniversalAddress();
    if (yield* this.createAta(t, i), yield* this.coreBridge.postVaa(t, r), n) {
      const d = new k(await this.getWrappedNative());
      if (Qt.equals(i.toUint8Array(), d.toUint8Array())) {
        yield* this.redeemAndUnwrap(t, r);
        return;
      }
    }
    const a = new k(t).unwrap(), o = r.payload.token.chain == this.chain ? ot : ur, s = new L().add(o(this.connection, this.tokenBridge.programId, this.coreBridge.address, a, r));
    s.feePayer = a, yield this.createUnsignedTx({ transaction: s }, "Solana.RedeemTransfer");
  }
  createUnsignedTx(t, r, n = !1) {
    return new ft(t, this.network, this.chain, r, n);
  }
}
const Cr = {
  Testnet: [
    "7VPWjBhCXrpYYBiRKZh1ubh9tLZZNkZGp2ReRphEV4Mc",
    "3WK3mEDNPrNuQReBvM28NcsqrExMnPxD9pPJmgrUeKKH",
    "BaGfF51MQ3a61papTRDYaNefBgTQ9ywnVne5fCff4bxT",
    "ACbmcQxbbhiXWM1GmapUSMmBYKMvnFLfAAXKqdo8xKwo",
    "GQtMXZxnuacCFTXVeTvyHi6P9F6chbtzhVc8JgD8hv7c",
    "3Ftc5hTz9sG4huk79onufGiebJNDMZNL8HYgdMJ9E7JR",
    "DMw2tLaq1bVoAEKtkoUtieSk9bfCPUvubYLatTMsSVop",
    "84F2QX9278ToDmA98u4A86xSV9hz1ovazr8zwGaX6qjS",
    "So11111111111111111111111111111111111111112",
    "8987WGkYa5viiZ9DD8sS3PB5XghKmWjkEgmzvwDuoAEc",
    "BJZ72CjPQojVoH68mzrd4VQ4nr6KuhbAGnhZEZCujKxY"
  ],
  Mainnet: [
    "7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs",
    "A9mUU4qviSctJVPJdBJWkb28deg915LYJKrzQ19ji3FM",
    "Dn4noZ5jgGfkntzcQSUZ8czkreiZ1ForXYoV2H8Dm7S1",
    "3NZ9JMVBmGAqocybic2c7LQCJScmgsAZ6vQqTDzcqmJh",
    "9gP2kCy3wA1ctvYWQk75guqXuHfrEomqydHLtcTCqiLa",
    "Gz7VkD4MacbEB6yC5XD3HcumEiYx2EtDYYrfikGsvopG",
    "KgV1GvrHQmRBY8sHQQeUKwTm2r2h8t4C8qt12Cw1HVE",
    "DRQBDBEWmwWGK13fRTLhSPzjbvMSUavhV6nW4RUH8W6T",
    "9kvAcwQbqejuJMd59mKuw2bfSsLRaQ7zuvaTVHEeBBec",
    "7ixSaXGsHAFy34wogPk2YXiUX3BMmQMFdercdaHLnBby",
    "G1vJEgzepqhnVu35BN4jrkv3wVwkujYWFFCxhbEZ1CZr",
    "So11111111111111111111111111111111111111112"
  ],
  Devnet: []
}, Q = 9, se = new v(10), J = new v(1e8);
class Ke {
  constructor(t, r, n, i) {
    f(this, "network");
    f(this, "chain");
    f(this, "connection");
    f(this, "contracts");
    f(this, "chainId");
    f(this, "coreBridgeProgramId");
    f(this, "tokenBridgeProgramId");
    f(this, "tokenBridgeRelayer");
    this.network = t, this.chain = r, this.connection = n, this.contracts = i, this.chainId = y(r);
    const a = i.tokenBridgeRelayer;
    if (!a)
      throw new Error(`TokenBridge contract Address for chain ${r} not found`);
    this.tokenBridgeRelayer = De(a, n), this.tokenBridgeProgramId = new c(i.tokenBridge), this.coreBridgeProgramId = new c(i.coreBridge);
  }
  static async fromRpc(t, r) {
    const [n, i] = await oe.chainFromRpc(t), a = r[i];
    if (a.network !== n)
      throw new Error(`Network mismatch for chain ${i}: ${a.network} != ${n}`);
    return new Ke(n, i, t, a.contracts);
  }
  async *transfer(t, r, n, i, a) {
    const s = new k(t).unwrap(), d = r.address.toUniversalAddress().toUint8Array(), u = this.mintAddress(n), m = new L();
    if ($(n)) {
      const p = We(u, s);
      try {
        await tn(this.connection, p);
      } catch (C) {
        if (C instanceof Re) {
          const R = At(s, p, s, u);
          m.add(R);
        } else
          throw C;
      }
    }
    const h = a || 0n, g = $(n), w = g ? await br(this.connection, this.tokenBridgeRelayer.programId, s, this.tokenBridgeProgramId, this.coreBridgeProgramId, u, i, h, d, r.chain, 0, g) : await Tr(this.connection, this.tokenBridgeRelayer.programId, s, this.tokenBridgeProgramId, this.coreBridgeProgramId, u, i, h, d, r.chain, 0);
    m.add(w), m.feePayer = s, yield this.createUnsignedTx({ transaction: m }, "AutomaticTokenBridge.Transfer");
  }
  async *redeem(t, r) {
    const n = new L();
    throw yield this.createUnsignedTx({ transaction: n }, "AutomaticTokenBridge.Redeem"), new Error("Method not implemented.");
  }
  async getRelayerFee(t, r) {
    const n = this.mintAddress(r), [{ fee: i }, { swapRate: a }, { relayerFeePrecision: o }] = await Promise.all([
      this.getForeignContract(t),
      this.getRegisteredToken(n),
      this.getRedeemerConfig()
    ]), s = Number(await oe.getDecimals(this.chain, this.connection, r)), d = se.pow(new v(s)).mul(i).mul(J).div(new v(o).mul(a));
    return BigInt(d.toString());
  }
  async maxSwapAmount(t) {
    const r = this.mintAddress(t), [{ swapRate: n, maxNativeSwapAmount: i }, { swapRate: a }] = await Promise.all([
      this.getRegisteredToken(r),
      this.getRegisteredToken(D)
    ]), o = Number(await oe.getDecimals(this.chain, this.connection, t)), s = this.calculateNativeSwapRate(a, n), d = o > Q ? i.mul(s).mul(se.pow(new v(o - Q))).div(J) : i.mul(s).div(se.pow(new v(Q - o)).mul(J));
    return BigInt(d.toString());
  }
  async nativeTokenAmount(t, r) {
    if (r === 0n)
      return 0n;
    const n = this.mintAddress(t), i = Number(await oe.getDecimals(this.chain, this.connection, t)), [{ swapRate: a }, { swapRate: o }] = await Promise.all([
      this.getRegisteredToken(n),
      this.getRegisteredToken(D)
    ]), s = this.calculateNativeSwapRate(o, a), d = i > Q ? J.mul(new v(r.toString())).div(s.mul(se.pow(new v(i - Q)))) : J.mul(new v(r.toString())).mul(se.pow(new v(Q - i))).div(s);
    return BigInt(d.toString());
  }
  async isRegisteredToken(t) {
    var n;
    const r = this.mintAddress(t);
    try {
      return await this.getRegisteredToken(r), !0;
    } catch (i) {
      if ((n = i.message) != null && n.includes("Account does not exist"))
        return !1;
      throw i;
    }
  }
  mintAddress(t) {
    return $(t) ? new c(D) : new k(t).unwrap();
  }
  async getRegisteredTokens() {
    return Cr[this.network].map((t) => Be(this.chain, t));
  }
  calculateNativeSwapRate(t, r) {
    return J.mul(t).div(r);
  }
  async getForeignContract(t) {
    return await this.tokenBridgeRelayer.account.foreignContract.fetch(Ie(this.tokenBridgeRelayer.programId, t));
  }
  async getRegisteredToken(t) {
    return await this.tokenBridgeRelayer.account.registeredToken.fetch(Ue(this.tokenBridgeRelayer.programId, t));
  }
  async getRedeemerConfig() {
    return await this.tokenBridgeRelayer.account.redeemerConfig.fetch(Dn(this.tokenBridgeRelayer.programId));
  }
  createUnsignedTx(t, r, n = !1) {
    return new ft(t, this.network, this.chain, r, n);
  }
}
mt(gt, "TokenBridge", qe);
mt(gt, "AutomaticTokenBridge", Ke);
export {
  pe as CreateMetadataAccountArgs,
  K as Creator,
  be as Data,
  Fe as EndpointRegistration,
  at as Key,
  Le as Metadata,
  Ke as SolanaAutomaticTokenBridge,
  qe as SolanaTokenBridge,
  de as SplTokenMetadataProgram,
  vt as TOKEN_BRIDGE_IDL,
  kr as TOKEN_BRIDGE_RELAYER_IDL,
  Ne as TokenBridgeConfig,
  Oe as WrappedMeta,
  sr as coder,
  it as createApproveAuthoritySignerInstruction,
  or as createAttestTokenInstruction,
  ot as createCompleteTransferNativeInstruction,
  ur as createCompleteTransferWrappedInstruction,
  lr as createCreateWrappedInstruction,
  Pr as createInitializeInstruction,
  W as createReadOnlyTokenBridgeProgramInterface,
  Wr as createRegisterChainInstruction,
  ar as createTokenBridgeProgramInterface,
  De as createTokenBridgeRelayerProgramInterface,
  ct as createTransferNativeInstruction,
  br as createTransferNativeTokensWithRelayInstruction,
  ut as createTransferNativeWithPayloadInstruction,
  wr as createTransferWrappedInstruction,
  Tr as createTransferWrappedTokensWithRelayInstruction,
  Sr as createTransferWrappedWithPayloadInstruction,
  Ir as createUpgradeContractInstruction,
  ne as deriveAuthoritySignerKey,
  Me as deriveCustodyKey,
  fe as deriveCustodySignerKey,
  te as deriveEndpointKey,
  Ie as deriveForeignContractAddress,
  ke as deriveMintAuthorityKey,
  ze as deriveRedeemerAccountKey,
  Dn as deriveRedeemerConfigAddress,
  Ue as deriveRegisteredTokenAddress,
  xe as deriveSenderAccountKey,
  Mt as deriveSenderConfigAddress,
  Te as deriveSplTokenMetadataKey,
  kt as deriveTmpTokenAccountAddress,
  B as deriveTokenBridgeConfigKey,
  bt as deriveTokenTransferMessageAddress,
  Y as deriveWrappedMetaKey,
  re as deriveWrappedMintKey,
  st as getAttestTokenAccounts,
  cr as getCompleteTransferNativeAccounts,
  Nr as getCompleteTransferNativeWithPayloadCpiAccounts,
  dr as getCompleteTransferWrappedAccounts,
  Fr as getCompleteTransferWrappedWithPayloadCpiAccounts,
  fr as getCreateWrappedAccounts,
  Er as getEndpointRegistration,
  mr as getInitializeAccounts,
  _r as getMetadata,
  gr as getRegisterChainAccounts,
  Rr as getTokenBridgeConfig,
  Ur as getTokenBridgeDerivedAccounts,
  pr as getTransferNativeAccounts,
  Bt as getTransferNativeWithPayloadAccounts,
  Ar as getTransferNativeWithPayloadCpiAccounts,
  yr as getTransferWrappedAccounts,
  Rt as getTransferWrappedWithPayloadAccounts,
  Mr as getTransferWrappedWithPayloadCpiAccounts,
  hr as getUpgradeContractAccounts,
  me as getWrappedMeta
};
