// @wormhole-foundation/wormhole-connect/lib/wallets-B2tqS5lZ.mjs
var p = function(e, t, r, n) {
  if (r === "a" && !n) throw new TypeError("Private accessor was defined without a getter");
  if (typeof t == "function" ? e !== t || !n : !t.has(e)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
  return r === "m" ? n : r === "a" ? n.call(e) : n ? n.value : t.get(e);
}, w = function(e, t, r, n, a) {
  if (n === "m") throw new TypeError("Private method is not writable");
  if (n === "a" && !a) throw new TypeError("Private accessor was defined without a setter");
  if (typeof t == "function" ? e !== t || !a : !t.has(e)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
  return n === "a" ? a.call(e, r) : a ? a.value = r : t.set(e, r), r;
}, s;
let d;
const c = /* @__PURE__ */ new Set();
function h(e) {
  i = void 0, c.add(e);
}
function g(e) {
  i = void 0, c.delete(e);
}
const o = {};
function m() {
  if (d || (d = Object.freeze({ register: l, get: y, on: E }), typeof window > "u"))
    return d;
  const e = Object.freeze({ register: l });
  try {
    window.addEventListener("wallet-standard:register-wallet", ({ detail: t }) => t(e));
  } catch (t) {
    console.error(`wallet-standard:register-wallet event listener could not be added
`, t);
  }
  try {
    window.dispatchEvent(new b(e));
  } catch (t) {
    console.error(`wallet-standard:app-ready event could not be dispatched
`, t);
  }
  return d;
}
function l(...e) {
  var t;
  return e = e.filter((r) => !c.has(r)), e.length ? (e.forEach((r) => h(r)), (t = o.register) == null || t.forEach((r) => f(() => r(...e))), function() {
    var n;
    e.forEach((a) => g(a)), (n = o.unregister) == null || n.forEach((a) => f(() => a(...e)));
  }) : () => {
  };
}
let i;
function y() {
  return i || (i = [...c]), i;
}
function E(e, t) {
  var r;
  return (r = o[e]) != null && r.push(t) || (o[e] = [t]), function() {
    var a;
    o[e] = (a = o[e]) == null ? void 0 : a.filter((u) => t !== u);
  };
}
function f(e) {
  try {
    e();
  } catch (t) {
    console.error(t);
  }
}
class b extends Event {
  get detail() {
    return p(this, s, "f");
  }
  get type() {
    return "wallet-standard:app-ready";
  }
  constructor(t) {
    super("wallet-standard:app-ready", {
      bubbles: !1,
      cancelable: !1,
      composed: !1
    }), s.set(this, void 0), w(this, s, t, "f");
  }
  /** @deprecated */
  preventDefault() {
    throw new Error("preventDefault cannot be called");
  }
  /** @deprecated */
  stopImmediatePropagation() {
    throw new Error("stopImmediatePropagation cannot be called");
  }
  /** @deprecated */
  stopPropagation() {
    throw new Error("stopPropagation cannot be called");
  }
}
s = /* @__PURE__ */ new WeakMap();
export {
  m as g
};
